{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Scorer Dev GH 22 01\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "****************************\n",
    "* consolidation of protein scoring routines into a single module.\n",
    "* 5 Sep 2018\n",
    "* separated out into a module to enable independent tesign and version control\n",
    "* 14 Sep 2018 include randomising by mask\n",
    "*\n",
    "* 22 Sep 2018\n",
    "* added randomseed to randomising routines\n",
    "* allows seed to be set externally and hence allows reproducibility where required.\n",
    "*\n",
    "* 8 Nov 2018\n",
    "* change the scoring of X as requested by Tim\n",
    "* currently \"X\" scores\n",
    "* 2\n",
    "* 4 2\n",
    "* 2 2\n",
    "* 2 2    across the 7 main residues. This should be changed to\n",
    "\n",
    "* 2\n",
    "* 4 2\n",
    "* 2 2\n",
    "* 0 2 \n",
    "* \n",
    "* and also add 'XXXXXX' on the front and 'XXXX' on the end\n",
    "* this done via an argument\n",
    "*\n",
    "* 13 Nov 2018\n",
    "* added method score_acidic_tract for just the specific acid tract prior to the FFAT motif.\n",
    "*\n",
    "* 23 May 2019\n",
    "* adding a new class for a separate scoring model - mail from Tim 21 May 2019\n",
    "* for Fabien Alpy\n",
    "* \"would like to see what the genome has for motifs with S/T at position 4 and no prejudice for S/T at all other positions.\"\n",
    "*\n",
    "* 22 April 2020\n",
    "* forced upgrade to python 3\n",
    "* print 'xxx' replaced with print ('xxx')\n",
    "* changes to lists and ranges showing up in random.shuddle) so replacing range(0,n) with list(range(0,n))\n",
    "* \n",
    "****************************\n",
    "'''\n",
    "\n",
    "import random # randomising sequences\n",
    "\n",
    "class SequenceAnalysis():\n",
    "    \n",
    "    def __init__(self):\n",
    "        '''\n",
    "        ****************************\n",
    "        * 4 Aug 2018\n",
    "        * change of weightings from previous version\n",
    "        * inclusion of * as an amino acid with its own set of weightings\n",
    "        * 10 Aug inclusion of U same weightings as K for human proteins\n",
    "        * \n",
    "        ****************************\n",
    "        '''\n",
    "    \n",
    "        self.versionid = 'Dev on 1.0'\n",
    "    \n",
    "        self.aminoacids = 'ACDEFGHIKLMNPQRSTUVWXY'\n",
    "        self.aminoacids += 'Z'# special because Z crops up in one proteome\n",
    "        #self.aminoacids += '*'# occurs in some proteomes so added here - 22 Sep 18\n",
    "\n",
    "        self.aminoscores= []\n",
    "        self.aminoscores.append(['DE',0])\n",
    "        self.aminoscores.append(['ST',0.5])\n",
    "        self.aminoscores.append(['ACFGHILMNPQVWY',1])\n",
    "        self.aminoscores.append(['KRUXZ*',2])\n",
    "\n",
    "        self.aminoacidgroups = []\n",
    "        self.aminoacidgroups.append([1,'DE'])\n",
    "        self.aminoacidgroups.append([2,'ST'])\n",
    "        self.aminoacidgroups.append([3,'FYW'])\n",
    "        self.aminoacidgroups.append([4,'UKRHXZ'])\n",
    "        self.aminoacidgroups.append([5,'ACG'])\n",
    "        self.aminoacidgroups.append([6,'ILMV'])\n",
    "        self.aminoacidgroups.append([7,'NQ'])\n",
    "        self.aminoacidgroups.append([8,'P'])\n",
    "\n",
    "        self.flankrule = [[0,0],[2.0,0.5],[3.0,1.0],[4.0,1.5]]\n",
    "\n",
    "        self.weightings = []\n",
    "\n",
    "        self.weightings.append(['*',[2,4,2,2,2,2,2]])\n",
    "        self.weightings.append(['A',[1,4,1,2,0,0,1]])\n",
    "        self.weightings.append(['C',[1,4,1,2,0,0,1]])\n",
    "        self.weightings.append(['D',[0,4,1,0,4,0,0]])\n",
    "        self.weightings.append(['E',[0,4,1,0,4,0,0]])\n",
    "        self.weightings.append(['F',[1,0,0,2,2,0,1]])\n",
    "        self.weightings.append(['G',[1,4,1,2,1,0,1]])\n",
    "        self.weightings.append(['H',[1,4,0.7,2,2,0,1.5]])\n",
    "        self.weightings.append(['I',[1,4,1,2,2,0,1]])\n",
    "        self.weightings.append(['K',[1.5,4,1,2,2.5,0,1.5]])\n",
    "        self.weightings.append(['L',[1,4,1,2,2,0,1]])\n",
    "        self.weightings.append(['M',[1,4,1,2,2,0,1]])\n",
    "        self.weightings.append(['N',[1,4,1,2,2,0,1]])\n",
    "        self.weightings.append(['P',[1,4,1,2,2,0,1]])\n",
    "        self.weightings.append(['Q',[1,4,1,2,2,0,1]])\n",
    "        self.weightings.append(['R',[2,4,1,2,3,0,1.5]])\n",
    "        self.weightings.append(['S',[0.5,4,1,0.5,0.5,0,0.5]])\n",
    "        self.weightings.append(['T',[0.5,4,1,0.5,0.5,0,0.5]])\n",
    "        self.weightings.append(['U',[1.5,4,1,2,2.5,0,1.5]])#10 Aug for human protein\n",
    "        self.weightings.append(['V',[1,4,1,2,2,0,1]])\n",
    "        self.weightings.append(['W',[1,2,0.7,2,2,0,1]])\n",
    "        self.weightings.append(['Y',[1,0, 0.5,2,2,0,1]])\n",
    "        self.weightings.append(['X',[2,4,2,2,2,0,2]])\n",
    "        self.weightings.append(['Z',[2,4,2,2,2,2,2]])\n",
    "        \n",
    "    def score_acidic_tract(self, sequence):\n",
    "        \n",
    "        '''\n",
    "        ***********************************\n",
    "        * 13 Nov 2018\n",
    "        * scores the acidic tract separately\n",
    "        * currently does not check length of sequence\n",
    "        ***********************************\n",
    "        '''\n",
    "        \n",
    "        counttract = 0\n",
    "        myscore = 99\n",
    "        \n",
    "        for i in range(0,6):\n",
    "            for score in self.aminoscores:\n",
    "                if sequence[i] in score[0]:\n",
    "                    counttract += score[1]\n",
    "                    \n",
    "        for flankscore in self.flankrule:\n",
    "            if counttract > flankscore[0]:\n",
    "                myscore = flankscore[1]\n",
    "               \n",
    "        return myscore\n",
    "\n",
    "    def score_sequence(self, sequence):\n",
    "        \n",
    "        '''\n",
    "        ***********************************\n",
    "        * applies the FFAT weighting algorithm to a sequence.\n",
    "        ***********************************\n",
    "        '''\n",
    "        myscore = 0\n",
    "\n",
    "        if len(sequence) < 13:\n",
    "            return -1\n",
    "        \n",
    "        countflank = 0.0\n",
    "        for i in range(0, 6):\n",
    "            for score in self.aminoscores:\n",
    "                if sequence[i] in score[0]:\n",
    "                    countflank += score[1]\n",
    "        \n",
    "        for flankscore in self.flankrule:\n",
    "            if countflank > flankscore[0]:\n",
    "                myscore = flankscore[1]\n",
    "\n",
    "        for i in range(6,13):\n",
    "            \n",
    "            idx = [x[0] for x in self.weightings].index(sequence[i])\n",
    "            weight = self.weightings[idx]\n",
    "            myscore += weight[1][i - 6]\n",
    "        \n",
    "        return myscore\n",
    "    \n",
    "    def score_protein(self, protein, bextended = True):\n",
    "        \n",
    "        '''\n",
    "        ***********************************\n",
    "        * returns the scores for a protein.\n",
    "        * currently does not add a string at the beginning to allow the first aminoacid string to be scored in the weightings\n",
    "        * nor pad out the last\n",
    "        *\n",
    "        * 8 Nov 2018 - pad out both ends with 'X' 6 at the beginning and 4 at the end.\n",
    "        * \n",
    "        ***********************************\n",
    "        '''\n",
    "        \n",
    "        if bextended:\n",
    "            myextendedprotein = 'XXXXXX' + protein[:] + 'XXXX'\n",
    "        else:\n",
    "            myextendedprotein = protein[:]\n",
    "    \n",
    "        nump = len(myextendedprotein)\n",
    "        myscores = []\n",
    "        for i in range(0,nump-14):\n",
    "            myscores.append(self.score_sequence(myextendedprotein[i : i + 13]))\n",
    "        \n",
    "        return myscores\n",
    "    \n",
    "    def score_protein_acidic_tract(self, protein):\n",
    "        \n",
    "        nump = len(protein)\n",
    "        myscores = []\n",
    "        for i in range(0, nump - 7):\n",
    "            myscores.append(self.score_acidic_tract(protein[i : i + 6]))\n",
    "        \n",
    "        return myscores\n",
    "    \n",
    "    def protein_score_ordered(self, rawscores, limit = 0):\n",
    "        \n",
    "        '''\n",
    "        ***********************************\n",
    "        * takes in the raw output of score_protein\n",
    "        * returns the scores sequenced from the lowest score to the highest\n",
    "        * score_protein returns a list of scores, eg [1.5, 3.4, 2.5, 0.5]\n",
    "        * this returns pairs, ie\n",
    "        * [[0.5, 3], [1.5, 0], [2.5, 2], [3.4, 1]]\n",
    "        *\n",
    "        * 17 Sep 2018\n",
    "        * sorting a 3000 length sequence when we just want the minimum 10 is a \n",
    "        * significant waste of CPU time so introduce an optional limit\n",
    "        * set to zero as default which means do the whole lot \n",
    "        ***********************************\n",
    "        '''\n",
    "        \n",
    "        lenN = len(rawscores)\n",
    "        seq = list(range(0,lenN))\n",
    "        \n",
    "        if limit > 0:\n",
    "            countlimit = min( limit + 1, lenN - 1)\n",
    "        else:\n",
    "            countlimit = lenN - 1\n",
    "            \n",
    "        for i in range (0, countlimit):\n",
    "            \n",
    "            for j in range(lenN - 2, i-1, -1):\n",
    "                \n",
    "                if rawscores[seq[j]] > rawscores[seq[j+1]]:\n",
    "                    temp = seq[j]\n",
    "                    seq[j] = seq[j+1]\n",
    "                    seq[j+1] = temp\n",
    "        \n",
    "        myorderedscores = []\n",
    "        \n",
    "        for i in range (0,lenN):\n",
    "            myorderedscores.append([rawscores[seq[i]], seq[i]])\n",
    "            \n",
    "        for score in myorderedscores:\n",
    "            if score[1] < 0:\n",
    "                print (score)\n",
    "            \n",
    "        return myorderedscores\n",
    "    \n",
    "    \n",
    "    def randomise_proteins_masked(self, proteins, mask, randomseed, bbyprotein = False):\n",
    "        '''\n",
    "        ***********************************\n",
    "        * 1 Sep 2018\n",
    "        * takes a list of proteins in and produces a randomised version\n",
    "        * stores the numbers of each amino acids\n",
    "        * produces a sequence of proteins of the same number and lengths as the incoming file\n",
    "        * and same number of amino acids in a random sequence\n",
    "        * 10 Sep 2018\n",
    "        * masked version. \n",
    "        * the mask is a string of 0's and 1's. Metaphorically the proteins are like a giant scrabble board\n",
    "        * a 1 in the mask means we keep that tile in its place. 0 means we pick it up\n",
    "        * put it in the bag, mix it up with the others, and put it down in a random place.\n",
    "        * proteome is [[file details],[proteins]]\n",
    "        * and protines are [[protein 1 name and details],[protein code]]\n",
    "        * \n",
    "        * 22 Sep 2018\n",
    "        * random seed added. Passed in externally\n",
    "        * change in randomisation to ensure same count of amino acids\n",
    "        ***********************************\n",
    "        '''\n",
    "        \n",
    "        random.seed(randomseed)\n",
    "        \n",
    "        numacids = len(self.aminoacids)   \n",
    "        \n",
    "        if bbyprotein:\n",
    "            \n",
    "            myrandomproteome = []\n",
    "            \n",
    "            numproteins = len(proteins)\n",
    "            for i in range(0, numproteins):\n",
    "                \n",
    "                lenprotein = len(proteins[i][1])\n",
    "                ctracids = lenprotein - sum(mask[i])\n",
    "                \n",
    "                # create  string of the amino acids in this protein\n",
    "                # only consider those that are to be shuffled i.e. mask == 0\n",
    "\n",
    "                acidstomix = ''\n",
    "            \n",
    "                for j in range(0, lenprotein):\n",
    "                    if mask[i][j] == 0:\n",
    "                        acidstomix += proteins[i][1][j]\n",
    "\n",
    "                # create massive string of the amino acids in same number\n",
    "\n",
    "                selectionsequence = ''\n",
    "                for acid in self.aminoacids:\n",
    "                    acidcount = acidstomix.count(acid) \n",
    "                    selectionsequence += acid * acidcount\n",
    "\n",
    "                #some proteins can contain acids not in our list.\n",
    "\n",
    "                nonacids = ''\n",
    "                for aa in acidstomix:\n",
    "                    if not aa in self.aminoacids:\n",
    "                        nonacids += aa\n",
    "\n",
    "                if len(nonacids)>0:\n",
    "                    print ('the following non-acids found: ' + nonacids)\n",
    "                    selectionsequence += nonacids\n",
    "\n",
    "\n",
    "                # create a randomised index aaindex which we will use sequentially.\n",
    "                # using random.shuffle!\n",
    "                # and use ctrindex to plough through\n",
    "\n",
    "                numchain = len(selectionsequence)\n",
    "                aaindex = list(range(0, numchain))\n",
    "                random.shuffle(aaindex)\n",
    "                ctrindex = 0\n",
    "\n",
    "                # programmer inefficiency here\n",
    "                # numchain, ctracids should be the same\n",
    "\n",
    "                newID = proteins[i][0][0] + ' - randomised'\n",
    "                thischain = ''\n",
    "                ctrstatic = 0\n",
    "\n",
    "                for j in range(0, lenprotein):\n",
    "\n",
    "                    if mask[i][j] == 0:\n",
    "\n",
    "                        thischain += selectionsequence[aaindex[ctrindex]]\n",
    "                        ctrindex += 1\n",
    "\n",
    "                    else:\n",
    "\n",
    "                        ctrstatic +=1\n",
    "                        thischain += proteins[i][1][j]\n",
    "\n",
    "                myrandomproteinheader = [newID,'static amino acids - ' + str(ctrstatic)]\n",
    "                myrandomprotein = [myrandomproteinheader, thischain]\n",
    "                myrandomproteome.append(myrandomprotein)\n",
    "            \n",
    "            \n",
    "        else:\n",
    "            \n",
    "             \n",
    "            aacount = []\n",
    "\n",
    "            ctracids = 0\n",
    "            numproteins = len(proteins)\n",
    "            for i in range(0, numproteins):\n",
    "                ctracids += len(proteins[i][1])\n",
    "                ctracids -= sum(mask[i]) # uses the fact that the mask is just 1's and 0's\n",
    "\n",
    "            # create massive string of the amino acids in the proteome\n",
    "            # only consider those that are to be shuffled i.e. mask == 0\n",
    "\n",
    "            acidstomix = ''\n",
    "            for i in range(0, numproteins):\n",
    "                numacidsinchain = len(proteins[i][1])\n",
    "                for j in range(0, numacidsinchain):\n",
    "                    if mask[i][j] == 0:\n",
    "                        acidstomix += proteins[i][1][j]\n",
    "\n",
    "            # create massive string of the amino acids in same number\n",
    "\n",
    "            selectionsequence = ''\n",
    "            for acid in self.aminoacids:\n",
    "                acidcount = acidstomix.count(acid) \n",
    "                selectionsequence += acid * acidcount\n",
    "\n",
    "            #some proteins can contain acids not in our list.\n",
    "\n",
    "            nonacids = ''\n",
    "            for aa in acidstomix:\n",
    "                if not aa in self.aminoacids:\n",
    "                    nonacids += aa\n",
    "\n",
    "            if len(nonacids)>0:\n",
    "                print ('the following non-acids found: ' + nonacids)\n",
    "                selectionsequence += nonacids\n",
    "\n",
    "\n",
    "            # create a randomised index aaindex which we will use sequentially.\n",
    "            # using random.shuffle!\n",
    "            # and use ctrindex to plough through\n",
    "\n",
    "            numchain = len(selectionsequence)\n",
    "            aaindex = list(range(0, numchain))\n",
    "            random.shuffle(aaindex)\n",
    "            ctrindex = 0\n",
    "\n",
    "            # programmer inefficiency here\n",
    "            # numchain, ctracids should be the same\n",
    "\n",
    "            myrandomproteome = []\n",
    "\n",
    "            for iprotein in range(0, numproteins):\n",
    "\n",
    "                lenthisprotein = len(proteins[iprotein][1])\n",
    "                newID = proteins[iprotein][0][0] + ' - randomised'\n",
    "                thischain = ''\n",
    "                ctrstatic = 0\n",
    "\n",
    "                for j in range(0, lenthisprotein):\n",
    "\n",
    "                    if mask[iprotein][j] == 0:\n",
    "\n",
    "                        thischain += selectionsequence[aaindex[ctrindex]]\n",
    "                        ctrindex += 1\n",
    "\n",
    "                    else:\n",
    "\n",
    "                        ctrstatic +=1\n",
    "                        thischain += proteins[iprotein][1][j]\n",
    "\n",
    "                myrandomproteinheader = [newID,'static amino acids - ' + str(ctrstatic)]\n",
    "                myrandomprotein = [myrandomproteinheader, thischain]\n",
    "                myrandomproteome.append(myrandomprotein)\n",
    "                            \n",
    "        return myrandomproteome\n",
    "    \n",
    "    def randomise_proteins(self, proteins, randomseed, bbyprotein = False):\n",
    "        '''\n",
    "        ***********************************\n",
    "        * 1 Sep 2018\n",
    "        * takes a proteome in and produces a randomised version\n",
    "        * stores the numbers of each amino acids\n",
    "        * produces a sequence of proteins of the same number and lengths as the incoming file\n",
    "        * and same number of amino acids in a random sequence\n",
    "        * 5 Sep 2018 modified to takew the full proteome structure.\n",
    "        * 14 Sep changed to use the masked version\n",
    "        ***********************************\n",
    "        '''\n",
    "        \n",
    "        blankmask = []\n",
    "\n",
    "        for protein in proteins:\n",
    "\n",
    "            numacids = len(protein[1])\n",
    "            proteinmask = [0] * numacids\n",
    "\n",
    "            blankmask.append(proteinmask)\n",
    "        \n",
    "        return self.randomise_proteins_masked(proteins,blankmask, randomseed, bbyprotein)\n",
    "    \n",
    "    def generate_mask_preserve_acidic_flank(self, proteins, maxscore = 0.5, bbyprotein = False):\n",
    "        '''\n",
    "        ***********************************\n",
    "        * 13 Dec 2018\n",
    "        * just produces a mask\n",
    "        * takes a proteome and produces a randomised version\n",
    "        * preserves acidic aminoacids 'DEST in the flank, but\n",
    "        * only with a sequence scoring 0.5 or less\n",
    "        ***********************************\n",
    "        '''\n",
    "        \n",
    "        blankmask = []\n",
    "        ctracidsintract = 0\n",
    "        qualifyingacids = 'DEST'\n",
    "\n",
    "        for protein in proteins:\n",
    "\n",
    "            numacids = len(protein[1])\n",
    "            proteinmask = [0] * numacids\n",
    "            \n",
    "            myresults = self.score_protein_acidic_tract(protein[1])\n",
    "            for i in range(0, len(myresults)):\n",
    "                if myresults[i] <= maxscore:\n",
    "                    # eureka - have found a qualifying segment.\n",
    "                    # check the flank for scidic amino acids\n",
    "                    # flank is 0-6 of the segment\n",
    "                    for j in range(0, 6):\n",
    "                        if protein[1][i + j] in qualifyingacids:\n",
    "                            proteinmask[i + j] = 1\n",
    "                            ctracidsintract += 1\n",
    "                    \n",
    "            blankmask.append(proteinmask)\n",
    "            \n",
    "        return blankmask\n",
    "\n",
    "    def randomise_proteins_preserve_acidic_flank(self, proteins, randomseed, maxscore = 0.5, bbyprotein = False):\n",
    "        '''\n",
    "        ***********************************\n",
    "        * 13 Dec 2018\n",
    "        * refactored\n",
    "        ***********************************\n",
    "        '''\n",
    "\n",
    "        mask = self.generate_mask_preserve_acidic_flank(proteins, maxscore, bbyprotein)\n",
    "        \n",
    "        return self.randomise_proteins_masked(proteins, mask, randomseed, bbyprotein)\n",
    "    \n",
    "    def aminoacid_count(self, proteins):\n",
    "        \n",
    "        '''\n",
    "        ***********************************\n",
    "        * 16 Sep 2018\n",
    "        * a utility for counting the number of amino acids in a proteome\n",
    "        * a check when randomising, and who knows what else\n",
    "        ***********************************\n",
    "        '''\n",
    "        numacids = len(self.aminoacids)\n",
    "        aacount = [0] * numacids\n",
    "        \n",
    "        for protein in proteins:\n",
    "            for i in range(0,numacids):\n",
    "                aacount[i] += protein[1].count(self.aminoacids[i]) \n",
    "         \n",
    "        return [self.aminoacids, aacount]\n",
    "        \n",
    "    def score_sequence_reordered(self, sequence, reorder):\n",
    "        \n",
    "        '''\n",
    "        ***********************************\n",
    "        * 28 Aug 2018\n",
    "        * a variant of scoresequence\n",
    "        * a re-order is passed in, eg 3672154\n",
    "        * which has to be passed in as 2561043\n",
    "        * it needs to have all the integers between 0 and 6 once and once only.\n",
    "        * I do not check for this.\n",
    "        *\n",
    "        * the reorder applies to weightings\n",
    "        *  weightings.append(['*',[2,4,2,2,2,2,2]])\n",
    "        * so for amino acids in range (6-13), instead of multiplying by weight[1][i-6]\n",
    "        * we multiply by weight[1][reorder[i - 6]]\n",
    "        * where weight = self.weightings row\n",
    "        ***********************************\n",
    "        '''\n",
    "        \n",
    "        myscore = 0\n",
    "        #mytally = [sequence,[]]\n",
    "\n",
    "        if len(sequence)<13:\n",
    "            return -1\n",
    "        \n",
    "        countflank = 0.0\n",
    "        for i in range(0, 6):\n",
    "            for score in self.aminoscores:\n",
    "                if sequence[i] in score[0]:\n",
    "                    countflank += score[1]\n",
    "                    #mytally[1].append(score[1])\n",
    "        \n",
    "        for flankscore in self.flankrule:\n",
    "            if countflank > flankscore[0]:\n",
    "                myscore = flankscore[1]\n",
    "\n",
    "        for i in range(6,13):\n",
    "            \n",
    "            idx = [x[0] for x in self.weightings].index(sequence[i])\n",
    "            weight = self.weightings[idx]\n",
    "            myscore += weight[1][reorder[i - 6]]\n",
    "            #mytally[1].append([sequence[i],weight[1][i - 6]])\n",
    "        \n",
    "        return myscore\n",
    "    \n",
    "    \n",
    "    def score_protein_reordered(self, protein, reorder, bextended = True):\n",
    "    \n",
    "        '''\n",
    "        ***********************************\n",
    "        * 8 Nov 2018 - pad out both ends with 'X' 6 at the beginning and 4 at the end.\n",
    "        * \n",
    "        ***********************************\n",
    "        '''\n",
    "        if bextended:\n",
    "            myextendedprotein = 'XXXXXX' + protein[:] + 'XXXX'\n",
    "        else:\n",
    "            myextendedprotein = protein[:]\n",
    "            \n",
    "        nump = len(myextendedprotein)\n",
    "        myscores = []\n",
    "        for i in range(0,nump-14):\n",
    "            myscores.append(self.score_sequence_reordered(myextendedprotein[i : i + 13],reorder))\n",
    "        \n",
    "        return myscores\n",
    "    \n",
    "    def acidity_score_protein(self, protein, stublength = 6):\n",
    "        \n",
    "        '''\n",
    "        ***********************************\n",
    "        * an idea from Tim.\n",
    "        * measure the amount of acidity in a short section and return in historgram form.\n",
    "        * ED = 1\n",
    "        * ST 0.5\n",
    "        * KR -H -0.5\n",
    "        * Others 0\n",
    "        ***********************************\n",
    "        '''\n",
    "        myscore = []\n",
    "        acidscore = [['ED',1.0],['ST',0.5],['KRH',-0.5]]\n",
    "        \n",
    "        lenprotein = len(protein)\n",
    "        bhavescore = False\n",
    "        currentscore = 0\n",
    "        \n",
    "        for i in range(0, (1 + lenprotein - stublength)):\n",
    "            if not bhavescore:\n",
    "                for j in range(i, i + stublength):\n",
    "                    for acid in acidscore:\n",
    "                        if protein[j] in acid[0]:\n",
    "                            currentscore += acid[1]\n",
    "                myscore.append(currentscore)\n",
    "                bhavescore = True\n",
    "            else:\n",
    "                for acid in acidscore:\n",
    "                    if protein[i - 1] in acid[0]:\n",
    "                        currentscore -= acid[1]\n",
    "                    if protein[i + stublength - 1] in acid[0]:\n",
    "                        currentscore += acid[1]\n",
    "                myscore.append(currentscore)\n",
    "        \n",
    "        return myscore\n",
    "    \n",
    " \n",
    "'''\n",
    "****************************\n",
    "* end of the scorer class\n",
    "****************************\n",
    "'''\n",
    "print ('Scorer Dev GH 22 01')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
