{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "New activity log version 1.2 Candidate release\n",
      "0: 20190128 18:21. scorer version = Dev on 1.0, logger version = Dev on 1.0\n",
      "1: 20190128 18:21. random seed set to 1548699711\n",
      "2: 20190128 18:21. Regression test\n",
      "3: 20190128 18:21. Analyser version - ProteomeAnalyser -Dev on 1.0\n",
      "4: 20190128 18:21. Scorer version - SequenceAnalysis -Dev on 1.0\n",
      "5: 20190128 18:21. Reader version - ProteomeReader - 1.1 Candidate release\n",
      "6: 20190128 18:21. results in /Users/johnslee/Documents/Regression_Test_20190128.txt\n",
      "replaced 1 instance of a rogue amino acid in MTTB1\n",
      "replaced 1 instance of a rogue amino acid in MTBB1\n",
      "replaced 1 instance of a rogue amino acid in MTMB\n",
      "replaced 1 instance of a rogue amino acid in MTTB2\n",
      "replaced 1 instance of a rogue amino acid in MTBB2\n",
      "replaced 1 instance of a rogue amino acid in MTBB3\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "string index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-408cca6fe586>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m    519\u001b[0m     \u001b[0;32mprint\u001b[0m \u001b[0;34m'finished TestMaskingAndRandomisation'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    520\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 521\u001b[0;31m \u001b[0mRunSystemRegressionTest\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    522\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    523\u001b[0m \u001b[0;31m#TestMaskingAndRandomisation()\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-3-408cca6fe586>\u001b[0m in \u001b[0;36mRunSystemRegressionTest\u001b[0;34m()\u001b[0m\n\u001b[1;32m    275\u001b[0m     \u001b[0mfout\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Analyser - '\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mmyanalysis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mversionid\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    276\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 277\u001b[0;31m     \u001b[0mfileretrieveresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mFileRetrieveTests\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mbtestmode\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmyreader\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    278\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    279\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfileretrieveresults\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-3-408cca6fe586>\u001b[0m in \u001b[0;36mFileRetrieveTests\u001b[0;34m(testmode, reader)\u001b[0m\n\u001b[1;32m    188\u001b[0m     \u001b[0mtestlist\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilelist\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 190\u001b[0;31m     \u001b[0mproteomesfortest\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mreader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread_files\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtestlist\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    191\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    192\u001b[0m     \u001b[0mcomparisontestlist\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-3-409d9995a2c3>\u001b[0m in \u001b[0;36mread_files\u001b[0;34m(self, flist)\u001b[0m\n\u001b[1;32m    170\u001b[0m             \u001b[0mfullfilename\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfpath\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mmyfile\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mftype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    171\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mmyfile\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 172\u001b[0;31m                 \u001b[0mmyproteins\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadFASTA\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfullfilename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    173\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0mmyfile\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    174\u001b[0m                 \u001b[0mmyproteins\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadUniRef50\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfullfilename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-3-409d9995a2c3>\u001b[0m in \u001b[0;36mreadFASTA\u001b[0;34m(self, fname)\u001b[0m\n\u001b[1;32m    264\u001b[0m                     \u001b[0midxID\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'>'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    265\u001b[0m                     \u001b[0mproteinID\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 266\u001b[0;31m                     \u001b[0;32mwhile\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0midxID\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m<>\u001b[0m\u001b[0;34m' '\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0midxID\u001b[0m\u001b[0;34m<\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    267\u001b[0m                         \u001b[0mproteinID\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0midxID\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    268\u001b[0m                         \u001b[0midxID\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: string index out of range"
     ]
    }
   ],
   "source": [
    "'''\n",
    "****************************\n",
    "* Proteome Test Harness\n",
    "* \n",
    "* set of functions to test the various modules\n",
    "* associated with the Yeast/Proteome project\n",
    "* 14 Sep 2018\n",
    "****************************\n",
    "'''\n",
    "\n",
    "#import datetime - imported in Proteome Core\n",
    "#import random - this is imported in Proteome Scorer\n",
    "import csv\n",
    "\n",
    "filenamecore = 'Proteome Core Dev.ipynb'\n",
    "filenamescorer = 'Proteome Scorer Dev.ipynb'\n",
    "filenameanalyser = 'Proteome Analyser Dev.ipynb'\n",
    "\n",
    "%run 'Proteome Core Dev.ipynb'\n",
    "%run 'Proteome Scorer Dev.ipynb'\n",
    "%run 'Proteome Analyser Dev.ipynb'\n",
    "\n",
    "def strip_identifiers(lineinfile):\n",
    "    \n",
    "    idx1 = lineinfile.index(',')\n",
    "    ref1 = lineinfile[:idx1]\n",
    "    idx2 = lineinfile[idx1 + 1:].index(',')\n",
    "    ref2 = lineinfile[idx1 + 1: idx1 + idx2 + 1]\n",
    "    \n",
    "    return [ref1,ref2]\n",
    "\n",
    "def compare_csv_files(filename1, filename2):\n",
    "    \n",
    "    '''\n",
    "    ********************************\n",
    "    * a routine for comparing two files that should be the same\n",
    "    * hmmm ... quite difficult. Need to make some assumptions.\n",
    "    * Firstly, we need a file to compare, and a file to compare against, so we take f1 and note its differences to f2.\n",
    "    * to do a full two-way comparison need to do compare_csv_files(f1,f2) followed by compare_csv_files(f2,f1)\n",
    "    *\n",
    "    * approach as follows\n",
    "    * need a row identifier which is the first two columns\n",
    "    * seek against the opposite current row and the next. Need to keep both rows, as may have\n",
    "    * f1(r1,r2,r4) and f2(r1,r2,r3,r4), or f1(r1,r2,r3), f2(r1,r3) \n",
    "    * \n",
    "    ********************************\n",
    "    '''\n",
    "    \n",
    "    f1 = open(filename1, 'r')\n",
    "    f2 = open(filename2, 'r')\n",
    "    \n",
    "    regressionstatus = []\n",
    "    regressionstatus.append('regression of ' + filename1 + ' against ' + filename2)\n",
    "    \n",
    "    bpass = True\n",
    "    \n",
    "    idx1 = 0\n",
    "    \n",
    "    file2lines = []\n",
    "    file2matched = []\n",
    "    file2ref = []\n",
    "    \n",
    "    # okay - this is crap as it consumes memory\n",
    "    for line2 in f2:\n",
    "        file2lines.append(line2)\n",
    "        file2matched.append(-1)# -1 means we do not have a match\n",
    "        file2ref.append(strip_identifiers(line2))\n",
    "        \n",
    "    num2 = len(file2lines)\n",
    "    \n",
    "    for line1 in f1:\n",
    "        \n",
    "        idx1 +=1\n",
    "        \n",
    "        baseline = line1\n",
    "        ref1 = strip_identifiers(baseline)\n",
    "        bfoundline = False\n",
    "        match = -1\n",
    "        \n",
    "        # now try and find the same line in f2\n",
    "        for i in range(0, num2):\n",
    "            if file2matched[i] == -1:\n",
    "                if ref1 == file2ref[i]:\n",
    "                    file2matched[i] = idx1\n",
    "                    bfoundline = True\n",
    "                    imatch = i\n",
    "                    break\n",
    "                    \n",
    "        if not bfoundline:\n",
    "            bpass = False\n",
    "            regressionstatus.append('row ' + str(idx1) + ' not matched - ' + ref1[0] + ', ' + ref1[1])\n",
    "        else:\n",
    "            if baseline == file2lines[imatch]:\n",
    "                pass\n",
    "            else:\n",
    "                bpass = False\n",
    "                regressionstatus.append('row ' + str(idx1) + ' found match but contents do not match - ' + ref1[0] + ', ' + ref1[1])\n",
    "                \n",
    "    return [bpass, regressionstatus]\n",
    "\n",
    "def readproteomefile(filedetails):\n",
    "    '''\n",
    "    ***************************\n",
    "    * is an independent check on the ability of ProteinReader() to properly read proteins\n",
    "    * this programme takes in text files of the various forms of proteomes\n",
    "    * and returns the number of proteins and the length of each protein.\n",
    "    * \n",
    "    * filedetails is a list of files in the form of the library conents ie [filename, reference name, readformat]\n",
    "    *\n",
    "    * some structure has to be assumed. the main one is that each new protein starts with a '>'\n",
    "    * slight buggeration factor. The 'dashed' format has lines such as '---A--' so need to remove the A's\n",
    "    ***************************\n",
    "    '''\n",
    "    \n",
    "    fpath = '/Users/johnslee/Documents/'\n",
    "    ftype = '.txt'\n",
    "    \n",
    "    # include all possible leters in aminoacidchars as do not want to terminate the read because of a wild card\n",
    "    aminoacidchars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ*'\n",
    "    \n",
    "    filemeasures = []\n",
    "    \n",
    "    for f in filedetails:\n",
    "        \n",
    "        proteincounts = []\n",
    "        \n",
    "        fname = fpath + f[0] + ftype\n",
    "        f = open(fname, 'r')\n",
    "        \n",
    "        thisprotein = ''\n",
    "        \n",
    "        for line in f:\n",
    "            \n",
    "            if '>' in line:\n",
    "                # new protein. May be the first in which case thisprotein is just ''\n",
    "\n",
    "                if len(thisprotein) > 0:\n",
    "                    \n",
    "                    thiscode = ''\n",
    "                    idx = len(thisprotein)\n",
    "\n",
    "                    bvalidcode = True\n",
    "                    while idx > 0 and bvalidcode:\n",
    "                        idx -= 1\n",
    "                        if thisprotein[idx] in aminoacidchars:\n",
    "                            thiscode = thisprotein[idx] + thiscode\n",
    "                        else:\n",
    "                            bvalidcode = thisprotein[idx] == '-' or thisprotein[idx] == chr(0)\n",
    "                            # hoping that there are no cases where - is in an annoying place other than for 'dashed' format\n",
    "                            # OrthoDB has a strange end character equal to chr(0)\n",
    "                            \n",
    "                    proteincounts.append(len(thiscode))\n",
    "                \n",
    "                thisprotein = ''\n",
    "            \n",
    "            else:\n",
    "                \n",
    "                thissegment = line.strip('\\n')\n",
    "                thissegment = thissegment.strip(chr(0))\n",
    "                thisprotein += thissegment.strip('-')\n",
    "       \n",
    "        thiscode = ''\n",
    "        for char in thisprotein:\n",
    "            if char in aminoacidchars:\n",
    "                thiscode += char\n",
    "                \n",
    "        proteincounts.append(len(thiscode))\n",
    "        \n",
    "        filemeasures.append([fname,len(proteincounts),proteincounts])\n",
    "    \n",
    "    return filemeasures\n",
    "        \n",
    "def FileRetrieveTests(testmode, reader):\n",
    "\n",
    "    '''\n",
    "    ***************************\n",
    "    * runs comparisons on a filelist between the formal file retrieval processes in Proteinreader\n",
    "    * and a basic text file reader with a minimum of assumptions.\n",
    "    * logger passed in to enable recording of the results\n",
    "    ***************************\n",
    "    '''\n",
    "\n",
    "    strresults = []\n",
    "    strresults.append('File retrieve comparison test - Core module against basic file retrieval routine.')\n",
    "    strresults.append(' ')\n",
    "    filelist = reader.availableproteomes()\n",
    "    \n",
    "    testlist = range(0,len(filelist))\n",
    "    \n",
    "    proteomesfortest = reader.read_files(testlist)\n",
    "    \n",
    "    comparisontestlist = []\n",
    "    for t in testlist:\n",
    "        comparisontestlist.append(filelist[t])\n",
    "        \n",
    "    comparisonproteomes = readproteomefile(comparisontestlist)\n",
    "    \n",
    "    # let the comparisons begin.\n",
    "    # assume number of proteomes is the same.\n",
    "    \n",
    "    numproteomes = len(proteomesfortest)\n",
    "    bpass = True\n",
    "    \n",
    "    for i in range(0,numproteomes):\n",
    "        \n",
    "        thisproteome = proteomesfortest[i][1]\n",
    "        checkproteome = comparisonproteomes[i]\n",
    "        \n",
    "        \n",
    "        strtext = filelist[testlist[i]][0]\n",
    "        strtext +=  ' number of proteins:'\n",
    "        strtext += ' reader =  ' + str(len(thisproteome))\n",
    "        strtext += ' check = ' + str(checkproteome[1])\n",
    "        strresults.append(strtext)\n",
    "        print strtext\n",
    "        \n",
    "        for j in range(0, len(thisproteome)):\n",
    "            if j < checkproteome[1]:\n",
    "                if len(thisproteome[j][1]) <> checkproteome[2][j]:\n",
    "                    bpass = False\n",
    "                    strtext = filelist[testlist[i]][0] + ': protein length mismatch - ' + thisproteome[j][0][0] + ' - ' +str(j)\n",
    "                    strtext += ' ' + str(len(thisproteome[j][1])) \n",
    "                    strtext += ' vs ' + str(checkproteome[2][j]) + ' in test file'\n",
    "                    strresults.append(strtext)\n",
    "                    print strtext\n",
    "                   \n",
    "    if bpass:\n",
    "        strresults.append('File retrieval test passed')\n",
    "    else:\n",
    "        strresults.append('*** File retrieval test failed ***')\n",
    "        \n",
    "    return strresults\n",
    "    \n",
    "\n",
    "def RunSystemRegressionTest():\n",
    "    \n",
    "    '''\n",
    "    ***************************\n",
    "    * 11th Nov 2018\n",
    "    * Full regression test\n",
    "    ***************************\n",
    "    '''\n",
    "    \n",
    "    btestmode = False\n",
    "    bprintmode = True\n",
    "    \n",
    "    '''\n",
    "    ***************************\n",
    "    * There are four components that need the versions recording.\n",
    "    * Proteome Core, which reads files and has utilities, and is loaded separately\n",
    "    * Proteome Scorer, which produces FFAT scores for protein chains\n",
    "    * Proteome Analyser, which has the scorer passed to it, does high level analysis and generates files\n",
    "    ***************************\n",
    "    '''\n",
    "\n",
    "    mylogger = ProteomeActivityLogger(btestmode, bprintmode)\n",
    "    myreader = ProteomeReader(btestmode)\n",
    "    myscorer = SequenceAnalysis()\n",
    "    myanalysis = ProteomeAnalyser(mylogger, myscorer, btestmode)\n",
    "    \n",
    "    mylogger.write_log_entry('Regression test')\n",
    "    mylogger.write_log_entry('Analyser version - ProteomeAnalyser -' + myanalysis.versionid)\n",
    "    mylogger.write_log_entry('Scorer version - SequenceAnalysis -' + myscorer.versionid)\n",
    "    mylogger.write_log_entry('Reader version - ProteomeReader - ' + myreader.versionid)\n",
    "    \n",
    "    foutname = myanalysis.filepath + 'Regression_Test_'+  datestamp() + '.txt'\n",
    "    mylogger.write_log_entry('results in ' + foutname)\n",
    "    \n",
    "    fout = open(foutname,\"w\")\n",
    "    \n",
    "    fout.write('Full System Regression Test' + '\\n')\n",
    "    fout.write('components' + '\\n')\n",
    "    fout.write('Core - ' + myreader.versionid + '\\n')\n",
    "    fout.write('Scorer - ' + myscorer.versionid + '\\n')\n",
    "    fout.write('Analyser - ' + myanalysis.versionid + '\\n')\n",
    "    \n",
    "    fileretrieveresults = FileRetrieveTests(btestmode,myreader)\n",
    "    \n",
    "    for result in fileretrieveresults:\n",
    "        fout.write(result + '\\n')\n",
    "    fout.write('\\n')\n",
    "    \n",
    "    '''\n",
    "    ***************************\n",
    "    * now compare the csv files\n",
    "    ***************************\n",
    "    '''\n",
    "    \n",
    "    fcsv1 = myreader.fpath + 'uniprot_proteome_yeastx6049_fasta_rawscores_20181108.csv'\n",
    "    ftest1 = myreader.fpath + 'uniprot_proteome_yeastx6049_fasta_rawscores_' + datestamp() + '.csv'\n",
    "    fcsv2 = myreader.fpath + 'uniprot_proteome_yeastx6049_fasta_bottomten_sorted_20181108.csv'\n",
    "    ftest2= myreader.fpath + 'uniprot_proteome_yeastx6049_fasta_bottomten_sorted_' + datestamp() + '.csv'\n",
    "    myproteome = myreader.read_files([1])\n",
    "    \n",
    "    myanalysis.run_analysis(0,myproteome)\n",
    "    csvcompareresults = compare_csv_files(ftest1,fcsv1)\n",
    "    \n",
    "    fout.write('comparison of ' + fcsv1+ ' against ' + ftest1 + '\\n')\n",
    "    \n",
    "    if csvcompareresults[0]:\n",
    "        fout.write(' Test Passed'+ '\\n')\n",
    "    else:\n",
    "        for result in csvcompareresults[1]:\n",
    "            fout.write(result + '\\n')\n",
    "            \n",
    "    fout.write('\\n')\n",
    "        \n",
    "    myanalysis.run_analysis(2,myproteome)\n",
    "    csvcompareresults = compare_csv_files(ftest2,fcsv2)\n",
    "    \n",
    "    fout.write('comparison of ' + fcsv2+ ' against ' + ftest2 + '\\n')\n",
    "    \n",
    "    if csvcompareresults[0]:\n",
    "        fout.write(' Test Passed'+ '\\n')\n",
    "    else:\n",
    "        for result in csvcompareresults[1]:\n",
    "            fout.write(result + '\\n')\n",
    "            \n",
    "    fout.write('\\n')\n",
    "    \n",
    "    print 'finished RunSystemRegressionTest'\n",
    "    \n",
    "    \n",
    "def AcidTractScore():\n",
    "    '''\n",
    "    ***************************\n",
    "    * Generates a csv file of acid tract scores\n",
    "    * does it on a specific test file, # 8 in the test list\n",
    "    ***************************\n",
    "    '''\n",
    "    \n",
    "    btestmode = True\n",
    "    \n",
    "    mylogger = ProteomeActivityLogger(False, True)\n",
    "    myreader = ProteomeReader(btestmode)\n",
    "    myscorer = SequenceAnalysis()\n",
    "    #myanalysis = ProteomeAnalyser(mylogger, myscorer, btestmode)\n",
    "    \n",
    "    myproteomes = myreader.read_files([8])\n",
    "    \n",
    "    fpath = '/Users/johnslee/Documents/'\n",
    "    \n",
    "    foutname = fpath + 'acidic_tract_score_test' + datestamp() + '.csv'\n",
    "    fout = open(foutname,\"w\")\n",
    "    \n",
    "    cr = csv.writer(fout, delimiter=',',quoting=csv.QUOTE_ALL)\n",
    "    \n",
    "    for proteome in myproteomes:\n",
    "        \n",
    "        myproteins = proteome[1]\n",
    "        \n",
    "        for protein in myproteins:\n",
    "        \n",
    "            myresults = myscorer.score_protein_acidic_tract(protein[1])\n",
    "            \n",
    "            randomseed = 1\n",
    "            maxscore = 0.0\n",
    "            bbyprotein = True\n",
    "            \n",
    "            mymask0 = myscorer.generate_mask_preserve_acidic_flank( [protein], maxscore , bbyprotein)\n",
    "            randomproteins0 = myscorer.randomise_proteins_masked([protein], mymask0, randomseed, bbyprotein)\n",
    "            \n",
    "            randomseed = 1\n",
    "            maxscore = 0.5\n",
    "            \n",
    "            mymask05 = myscorer.generate_mask_preserve_acidic_flank( [protein], maxscore , bbyprotein)\n",
    "            randomproteins05 = myscorer.randomise_proteins_masked([protein], mymask05, randomseed, bbyprotein)\n",
    "            \n",
    "            mycsv = []\n",
    "            mycsv.append(protein[0][0])\n",
    "            mycsv.append(protein[0][1])\n",
    "            mycsv.append(len(protein[1]))\n",
    "            \n",
    "            cr.writerow(mycsv)\n",
    "            \n",
    "            for i in range(0, len(myresults)):\n",
    "                mycsv = []\n",
    "                mycsv.append(protein[1][i])\n",
    "                mycsv.append(myresults[i])\n",
    "                mycsv.append(mymask0[0][i])\n",
    "                mycsv.append(randomproteins0[0][1][i])\n",
    "                mycsv.append(mymask05[0][i])\n",
    "                mycsv.append(randomproteins05[0][1][i])\n",
    "                cr.writerow(mycsv)\n",
    "\n",
    "    print 'finished writing ' + foutname\n",
    "                \n",
    "    \n",
    "def TestMaskingAndRandomisation():\n",
    "    '''\n",
    "    ***************************\n",
    "    * Functional test on screening and randomisation of proteome and proteins\n",
    "    * check mask is fully observed\n",
    "    * check number of acids fully preserved\n",
    "    ***************************\n",
    "    '''\n",
    "    \n",
    "    randomseed = 1\n",
    "    ftest='uniprot_proteome_yeastx6049_fasta'\n",
    "    \n",
    "    btestmode = False\n",
    "    \n",
    "    mylogger = ProteomeActivityLogger(False, True)\n",
    "    myreader = ProteomeReader(btestmode)\n",
    "    myscorer = SequenceAnalysis()\n",
    "    #myanalysis = ProteomeAnalyser(mylogger, myscorer, btestmode)\n",
    "    \n",
    "    myproteomes = myreader.read_files([1])\n",
    "    \n",
    "    # test masking.\n",
    "    # use the letter A\n",
    "    \n",
    "    maska = []\n",
    "    \n",
    "    mylogger.write_log_entry('Functional test - masking and randomisation')\n",
    "    mylogger.write_log_entry('scorer version ' + myscorer.versionid)\n",
    "    mylogger.write_log_entry('reader version ' + myreader.versionid)\n",
    "    \n",
    "    proteins = myproteomes[0][1]\n",
    "\n",
    "    for protein in proteins:\n",
    "\n",
    "        numacids = len(protein[1])\n",
    "        proteinmask = [0] * numacids\n",
    "            \n",
    "        for i in range(0, numacids):\n",
    "            if protein[1][i] == 'A':\n",
    "                proteinmask[i] = 1\n",
    "                \n",
    "        maska.append(proteinmask)\n",
    "        \n",
    "    # have the mask\n",
    "    # now2 test for by-protein randomisationby protein\n",
    "    \n",
    "    newproteins = myscorer.randomise_proteins_masked(proteins, maska, randomseed, True)\n",
    "    \n",
    "    numproteins = len(newproteins)\n",
    "    \n",
    "    bfullpass = True\n",
    "    \n",
    "    for i in range(0, numproteins):\n",
    "\n",
    "        bpass = True\n",
    "        acidcount = [[0,0]] * len(myscorer.aminoacids)\n",
    "        \n",
    "        for j in range(0, len(myscorer.aminoacids)):\n",
    "            acidcount[j][0] = proteins[i][1].count(myscorer.aminoacids[j])\n",
    "            acidcount[j][1] = newproteins[i][1].count(myscorer.aminoacids[j])\n",
    "            if acidcount[j][0] <> acidcount[j][1]:\n",
    "                bpass = False\n",
    "                mylogger.write_log_entry('mismatch in number for acid ' + myscorer.aminoacids[j] + ' in protein ' + str(i))\n",
    "                \n",
    "        bfullpass = bfullpass and bpass\n",
    "        \n",
    "        bpass = True\n",
    "        numproteinsinchain = len(proteins[i][1])\n",
    "\n",
    "        for j in range(0, numproteinsinchain):\n",
    "            \n",
    "            if proteins[i][1][j] == 'A':\n",
    "                if newproteins[i][1][j] <> 'A':\n",
    "                    bpass = False\n",
    "                    mylogger.write_log_entry('test failure - A transformed when should have been masked')\n",
    "                    \n",
    "            else:\n",
    "                if newproteins[i][1][j] == 'A':\n",
    "                    bpass = False\n",
    "                    mylogger.write_log_entry('test failure - A present when should not have been present')\n",
    "                    \n",
    "            bfullpass = bfullpass and bpass\n",
    "            \n",
    "    if bfullpass:\n",
    "        mylogger.write_log_entry('mask and randomisation by protein - test passed')\n",
    "    else:\n",
    "        mylogger.write_log_entry('mask and randomisation by protein - test failed')\n",
    "        \n",
    "    # now whole proteome\n",
    "    newproteins = myscorer.randomise_proteins_masked(proteins, maska, randomseed, False)\n",
    "    \n",
    "    numproteins = len(newproteins)\n",
    "    \n",
    "    bfullpass = True\n",
    "    \n",
    "    acidcount = [[0,0]] * len(myscorer.aminoacids)\n",
    "    \n",
    "    for i in range(0, numproteins):\n",
    "\n",
    "        bpass = True\n",
    "        numproteinsinchain = len(proteins[i][1])\n",
    "        \n",
    "        for j in range(0, len(myscorer.aminoacids)):\n",
    "            acidcount[j][0] += proteins[i][1].count(myscorer.aminoacids[j])\n",
    "            acidcount[j][1] += newproteins[i][1].count(myscorer.aminoacids[j])\n",
    "           \n",
    "        for j in range(0, numproteinsinchain):\n",
    "            \n",
    "            if proteins[i][1][j] == 'A':\n",
    "                if newproteins[i][1][j] <> 'A':\n",
    "                    bpass = False\n",
    "                    mylogger.write_log_entry('test failure - A transformed when should have been masked')\n",
    "                    \n",
    "            else:\n",
    "                if newproteins[i][1][j] == 'A':\n",
    "                    bpass = False\n",
    "                    mylogger.write_log_entry('test failure - A present when should not have been present')\n",
    "                    \n",
    "            bfullpass = bfullpass and bpass\n",
    "            \n",
    "    for j in range(0, len(myscorer.aminoacids)):\n",
    "        if acidcount[j][0] <> acidcount[j][1]:\n",
    "            bfullpass = False\n",
    "            mylogger.write_log_entry('mismatch in number for acid ' + myscorer.aminoacids[j])\n",
    "            \n",
    "    if bfullpass:\n",
    "        mylogger.write_log_entry('mask and randomisation for whole proteome - test passed')\n",
    "    else:\n",
    "        mylogger.write_log_entry('mask and randomisation for whole proteome - test failed')\n",
    "            \n",
    "    print 'finished TestMaskingAndRandomisation'\n",
    "    \n",
    "RunSystemRegressionTest()\n",
    "\n",
    "#TestMaskingAndRandomisation()\n",
    " \n",
    "#AcidTractScore()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "[[0,0]] * 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
