{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enter 0 for live or 1 for test (anything else exits) 0\n",
      "New activity log version 1.2 Candidate release **\n",
      "0: 20190214 19:27. scorer version = Dev on 1.0, logger version = Dev on 1.0\n",
      "1: 20190214 19:27. random seed set to 1550172476\n",
      "reader version = 1.1 Candidate release\n",
      "scorer version = Dev on 1.0\n",
      "analyser version = Dev on 1.0\n",
      "2: 20190214 19:27. core reader version = 1.1 Candidate release\n",
      "available files\n",
      "---------------\n",
      "0 orf_trans\n",
      "1 uniprot_proteome_yeastx6049_fasta\n",
      "2 HumanProteomeFasta\n",
      "3 Plasmodium_3D7_proteome_fasta\n",
      "4 Albugo_proteome_fasta\n",
      "5 uniprot_Arabidopsis_proteome_fasta\n",
      "6 uniprot_C_elegans_proteome_fasta\n",
      "7 uniprot_Drosophila_proteome_fasta\n",
      "8 UP000025675_M_tuberculosis_proteome_x3978_fasta\n",
      "9 UP000008816_S_aureus_x2889_fasta\n",
      "10 UP000051952_Bodo_saltans_x18124_fasta\n",
      "11 UP000009168_Tetrahymena_proteome_x26976_fasta\n",
      "12 Ubr1x1331 ipr003126(NOTpf02617pf13764pf12937+ident05fasta\n",
      "13 Rqc1=Tcf25x545identity05fasta\n",
      "14 Mcm3x196identity05fasta\n",
      "15 Seg2_blits3_x23\n",
      "16 Spironucleus_proteome_AUP000018208fasta\n",
      "17 Ecoli_proteome_AUP000000625fasta\n",
      "18 Giardia_proteome_AUP000001548fasta\n",
      "19 Mycoplasma_proteome_AUP000239951fasta\n",
      "20 uniprot-proteome_Angomonas_deanei_fasta\n",
      "21 uniprotproteome_Haloterrigena_turkmenica_Archea_fasta\n",
      "22 Kri1spf05178pf12936identity05\n",
      "23 Bbc1smti1identity05\n",
      "24 YPR097Wspf12828+pf12825+identity05\n",
      "25 YLR149Cspf10313+ipr015943+identity05\n",
      "26 RFIPsallfip+rbd+identity05\n",
      "27 Ede1x224_uniref_uniprot3A28uba+22eh+dom2229fasta\n",
      "28 FormatSeq_HALTV727b3_x615v2\n",
      "29 FormatSeq_HALTV299b_x135v2\n",
      "30 FormatSeq_HALTV265b3_x245v2\n",
      "31 E_histolyticaATCC30459_7959fasta\n",
      "32 VPS13chorein_Nvps13_atg2_caduniref50x3867fasta\n",
      "33 Erb1OrthoDBEOG093704ZYx537_506_species\n",
      "34 Opi1OrthoDBx200 fungiEOG092C5A9Y\n",
      "35 Ost1OrthDBx353EOG091G05K5\n",
      "36 Rnf31uniref50x343ex833fasta\n",
      "37 Sfb3_uniref50x488fasta\n",
      "38 Asg7x29fromOrthoDBfasta\n",
      "39 ACBD5_proteins\n",
      "40 ACBP_superfamily_proteins\n",
      "41 uniprot_proteome_AUP000001974_Sulfolobus_solfataricusx2938_fasta\n",
      "42 uniprot_proteome_AUP000002199_Archaeoglobus_fulgidusx2399_fasta\n",
      "43 uniprot_proteome_AUP000000805_Methanocaldococcus_jannaschiix1732_fasta\n",
      "44 uniprot_proteome_AUP000000595_Methanosarcina_mazeix3303_fasta\n",
      "45 uniprot_proteome_AUP000000752_Pyrococcus_horikoshiix2077_fasta\n",
      "46 uniprot-proteome_AUP000001024_Thermoplasma acidophilumx1482_fasta\n",
      "47 proteome_Perkinsela_fasta\n",
      "48 proteome_Paratrypanosoma_confusum_fasta\n",
      "49 proteome_Euglena_gracilis_fasta\n",
      "50 Phytomonas_Hart1_AUP000053358_fasta\n",
      "51 Phytomonas_EM1_AUP000027080_fasta\n",
      "52 Leptomonas_seymouri_AUP000038009_fasta\n",
      "53 Strigomonas_AUP000015354_fasta\n",
      "54 Trypanosoma_equiperdum_AUP000195570_fasta\n",
      "55 proteome_08076_E_dispar_fasta\n",
      "56 proteome_014680_E_invadens_fasta\n",
      "57 proteome_06769_E_nuttalli_fasta\n",
      "58 Vps13_Amoebozoa_167\n",
      "59 uniref-uniprot_UBR4_OR_pf13764_507fasta\n",
      "60 Ubr1_eog092c01d2_nr90\n",
      "61 Vid27_EOG092C1KQX_nr90\n",
      "62 Osh3_eog092c0t4h_nr90\n",
      "63 Exo1_eog092c2bju_nr90\n",
      "64 Kri1_EOG092C3V2Y_nr90\n",
      "65 Ost1_EOG092C29MC_nr90\n",
      "66 Ede1_EOG092C055E_nr90\n",
      "67 Bbc1_EOG092C0BBJ_nr90\n",
      "68 Opi1_EOG092C5A9Y_nr90\n",
      "69 YPR097W_eog092c0e1b_nr90\n",
      "70 OSH2_eog092c0gj4_nr90\n",
      "71 ERB1_EOG092C1381_nr90\n",
      "72 ASG7_eog092c5rpj_nr90\n",
      "73 NOP2_EOG092C2BNS\n",
      "74 SKP2_EOG092C0MR8\n",
      "75 STE13_EOG092C0PZJ\n",
      "76 UBP10_EOG092C5R71\n",
      "77 MCM3_EOG092C0VUM\n",
      "78 WSC2_EOG092C5SG6\n",
      "79 SFB3_EOG092C0DKG\n",
      "80 PET123_EOG092C4SPI\n",
      "81 RQC1_EOG092C0TUL\n",
      "82 YOR238W_EOG092C3RPQ\n",
      "83 Kar2_EOG092C1VPN\n",
      "84 Ylr149c_EOG092C28\n",
      "85 TUB1_EOG092C2JRV\n",
      "86 NOP2_EOG091G0685\n",
      "87 ACS1_EOG092C0Z1N\n",
      "88 ECM13_EOG092C4UTV\n",
      "89 EFR3_EOG092C0OH2\n",
      "90 ENT3_EOG092C3ZA8\n",
      "91 MET4_EOG092C5FM7\n",
      "92 MGR3_EOG092C3NE2\n",
      "93 MSH6_EOG092C0EWT\n",
      "94 RKR1_EOG092C029X\n",
      "95 SXM1_EOG092C0CGO\n",
      "96 YHL026C_EOG092C2HBH\n",
      "97 AIM44_EOG092C5FAO\n",
      "98 DYN1_EOG092C00NZ\n",
      "99 EPO1_EOG092C0XF3\n",
      "100 KXD1_EOG092C5F1D\n",
      "101 MDV1_EOG092C172M\n",
      "102 MGA2_EOG092C06NJ\n",
      "103 NUM1_EOG092C01BH\n",
      "104 PBI1_EOG092C41LE\n",
      "105 SEC2_EOG092C3RPI\n",
      "106 SEC27_EOG092C10HM\n",
      "107 SEC63_EOG092C20BT\n",
      "108 SEC66_EOG092C58AK\n",
      "109 TOR1_EOG092C00HJ\n",
      "110 UBP1_EOG092C3TMU\n",
      "111 UBP2_EOG092C04TL\n",
      "112 VPS13_EOG092C0SXB\n",
      "113 YJL049W_EOG092C3Z1D\n",
      "114 YML003W_EOG092C00HJ\n",
      "115 YML002W_EOG092C087J\n",
      "116 YPT7_EOG092C5MDH\n",
      "117 YPT7_FUNGALx38\n",
      "118 EDE1muc_EOG092C055E\n",
      "119 Eps15_homologues391\n",
      "120 Opi1_blastx101\n",
      "121 Kri1_pfx2nr90_metazoa_x226\n",
      "122 Erb1_BOP1NT_NOT_EOG092C1381x904\n",
      "123 MCM3_28266497at2759_metazoa_nr50\n",
      "124 MCM3_28266497at2759_metazoa_nr90\n",
      "125 NOP2_plants_nr90\n",
      "126 Sec24C_nonfungi_nr90\n",
      "127 Sec24ABD_nonfungi_nr90\n",
      "128 TUBA_nonfungi_nr90\n",
      "129 Tcf25_metazoa_nr90\n",
      "130 Tcf25_viridiplantae_nr90\n",
      "131 Efr3_viridiplantae_nr90\n",
      "132 Efr3_metazoa_nr90\n",
      "133 Sxm1_all_nr90\n",
      "134 Msh6_Nterms\n",
      "135 Irr1_fungi_nr90\n",
      "136 Irr1_metazoa_nr90\n",
      "137 Vps13_Fungal_select_nr90\n",
      "138 CHMP7_all_nr90\n",
      "139 Kxd1_all_nr90\n",
      "140 Sec2_all_nr90\n",
      "141 bNOP2_nr50_n650_POG091H02E2\n",
      "142 KRI1_pfx2nr90_NOTmetazoa_x672\n",
      "143 USP36_all_nr90\n",
      "144 Vid27_domain_nonF_nr90\n",
      "145 Vid27_name_nonF_nr90\n",
      "146 Epsin_all_nr90\n",
      "147 Kri1_viridiplantae_nr90_x79\n",
      "148 Epsin_plants_nr90\n",
      "149 MSH6_nonFungi_nr50\n",
      " \n",
      "enter files to test (no commas) 148 149\n",
      " \n",
      "available analyses\n",
      "---------------\n",
      "0 Proteome Score Creates a csv file with a full list of FFAT scores for each protein\n",
      "1 Bottom Ten Creates a csv file the lowest 10 scores for each protein plus location\n",
      "2 Bottom Ten - Sorted Creates a csv file the lowest 10 scores for each protein, sorted by lowest score. Very slow\n",
      "3 Histogram Creates a csv file of a histogram of the FFAT scores of a proteome\n",
      "4 Histogram + randomised histogram (full proteome) a histogram of the FFAT scores of a proteome plus a randomised equivalent histogram (by full proteome)\n",
      "5 Histogram + randomised histogram (within protein) as 4, but randomised within protein\n",
      "6 Reordered Histogram a histogram of the FFAT scores of a proteome plus a reordered weighting equivalent histogram\n",
      "7 Histogram + randomised preserving acids (full proteome) score = 0.5 a histogram of the FFAT scores of a proteome plus a randomised equivalent preserving acids in low-scoring flanks score = 0.5\n",
      "8 Histogram + randomised preserving acids (within protein) score = 0.5 as 7, but randomised within protein\n",
      "9 Histogram + randomised preserving acids (full proteome) score = 0.0 a histogram of the FFAT scores of a proteome plus a randomised equivalent preserving acids in low-scoring flanks\n",
      "10 Histogram + randomised preserving acids (within protein) score = 0.0 as 7, but randomised within protein\n",
      "Enter analysis to test 2\n",
      " \n",
      "3: 20190214 19:28. running bottom_ten_sorted on proteome Epsin_plants_nr90\n",
      "4: 20190214 19:28. finished sorted and ranking proteins in Epsin_plants_nr90\n",
      "5: 20190214 19:28. running bottom_ten_sorted on proteome MSH6_nonFungi_nr50\n",
      "6: 20190214 19:28. finished sorted and ranking proteins in MSH6_nonFungi_nr50\n",
      "7: 20190214 19:28. finished bottom ten - sorted\n",
      " \n",
      "finished analysis\n",
      " \n",
      "Enter 0 for live or 1 for test (anything else exits) 0\n",
      "8: 20190214 19:28. scorer version = Dev on 1.0, logger version = Dev on 1.0\n",
      "9: 20190214 19:28. random seed set to 1550172537\n",
      "reader version = 1.1 Candidate release\n",
      "scorer version = Dev on 1.0\n",
      "analyser version = Dev on 1.0\n",
      "10: 20190214 19:28. core reader version = 1.1 Candidate release\n",
      "available files\n",
      "---------------\n",
      "0 orf_trans\n",
      "1 uniprot_proteome_yeastx6049_fasta\n",
      "2 HumanProteomeFasta\n",
      "3 Plasmodium_3D7_proteome_fasta\n",
      "4 Albugo_proteome_fasta\n",
      "5 uniprot_Arabidopsis_proteome_fasta\n",
      "6 uniprot_C_elegans_proteome_fasta\n",
      "7 uniprot_Drosophila_proteome_fasta\n",
      "8 UP000025675_M_tuberculosis_proteome_x3978_fasta\n",
      "9 UP000008816_S_aureus_x2889_fasta\n",
      "10 UP000051952_Bodo_saltans_x18124_fasta\n",
      "11 UP000009168_Tetrahymena_proteome_x26976_fasta\n",
      "12 Ubr1x1331 ipr003126(NOTpf02617pf13764pf12937+ident05fasta\n",
      "13 Rqc1=Tcf25x545identity05fasta\n",
      "14 Mcm3x196identity05fasta\n",
      "15 Seg2_blits3_x23\n",
      "16 Spironucleus_proteome_AUP000018208fasta\n",
      "17 Ecoli_proteome_AUP000000625fasta\n",
      "18 Giardia_proteome_AUP000001548fasta\n",
      "19 Mycoplasma_proteome_AUP000239951fasta\n",
      "20 uniprot-proteome_Angomonas_deanei_fasta\n",
      "21 uniprotproteome_Haloterrigena_turkmenica_Archea_fasta\n",
      "22 Kri1spf05178pf12936identity05\n",
      "23 Bbc1smti1identity05\n",
      "24 YPR097Wspf12828+pf12825+identity05\n",
      "25 YLR149Cspf10313+ipr015943+identity05\n",
      "26 RFIPsallfip+rbd+identity05\n",
      "27 Ede1x224_uniref_uniprot3A28uba+22eh+dom2229fasta\n",
      "28 FormatSeq_HALTV727b3_x615v2\n",
      "29 FormatSeq_HALTV299b_x135v2\n",
      "30 FormatSeq_HALTV265b3_x245v2\n",
      "31 E_histolyticaATCC30459_7959fasta\n",
      "32 VPS13chorein_Nvps13_atg2_caduniref50x3867fasta\n",
      "33 Erb1OrthoDBEOG093704ZYx537_506_species\n",
      "34 Opi1OrthoDBx200 fungiEOG092C5A9Y\n",
      "35 Ost1OrthDBx353EOG091G05K5\n",
      "36 Rnf31uniref50x343ex833fasta\n",
      "37 Sfb3_uniref50x488fasta\n",
      "38 Asg7x29fromOrthoDBfasta\n",
      "39 ACBD5_proteins\n",
      "40 ACBP_superfamily_proteins\n",
      "41 uniprot_proteome_AUP000001974_Sulfolobus_solfataricusx2938_fasta\n",
      "42 uniprot_proteome_AUP000002199_Archaeoglobus_fulgidusx2399_fasta\n",
      "43 uniprot_proteome_AUP000000805_Methanocaldococcus_jannaschiix1732_fasta\n",
      "44 uniprot_proteome_AUP000000595_Methanosarcina_mazeix3303_fasta\n",
      "45 uniprot_proteome_AUP000000752_Pyrococcus_horikoshiix2077_fasta\n",
      "46 uniprot-proteome_AUP000001024_Thermoplasma acidophilumx1482_fasta\n",
      "47 proteome_Perkinsela_fasta\n",
      "48 proteome_Paratrypanosoma_confusum_fasta\n",
      "49 proteome_Euglena_gracilis_fasta\n",
      "50 Phytomonas_Hart1_AUP000053358_fasta\n",
      "51 Phytomonas_EM1_AUP000027080_fasta\n",
      "52 Leptomonas_seymouri_AUP000038009_fasta\n",
      "53 Strigomonas_AUP000015354_fasta\n",
      "54 Trypanosoma_equiperdum_AUP000195570_fasta\n",
      "55 proteome_08076_E_dispar_fasta\n",
      "56 proteome_014680_E_invadens_fasta\n",
      "57 proteome_06769_E_nuttalli_fasta\n",
      "58 Vps13_Amoebozoa_167\n",
      "59 uniref-uniprot_UBR4_OR_pf13764_507fasta\n",
      "60 Ubr1_eog092c01d2_nr90\n",
      "61 Vid27_EOG092C1KQX_nr90\n",
      "62 Osh3_eog092c0t4h_nr90\n",
      "63 Exo1_eog092c2bju_nr90\n",
      "64 Kri1_EOG092C3V2Y_nr90\n",
      "65 Ost1_EOG092C29MC_nr90\n",
      "66 Ede1_EOG092C055E_nr90\n",
      "67 Bbc1_EOG092C0BBJ_nr90\n",
      "68 Opi1_EOG092C5A9Y_nr90\n",
      "69 YPR097W_eog092c0e1b_nr90\n",
      "70 OSH2_eog092c0gj4_nr90\n",
      "71 ERB1_EOG092C1381_nr90\n",
      "72 ASG7_eog092c5rpj_nr90\n",
      "73 NOP2_EOG092C2BNS\n",
      "74 SKP2_EOG092C0MR8\n",
      "75 STE13_EOG092C0PZJ\n",
      "76 UBP10_EOG092C5R71\n",
      "77 MCM3_EOG092C0VUM\n",
      "78 WSC2_EOG092C5SG6\n",
      "79 SFB3_EOG092C0DKG\n",
      "80 PET123_EOG092C4SPI\n",
      "81 RQC1_EOG092C0TUL\n",
      "82 YOR238W_EOG092C3RPQ\n",
      "83 Kar2_EOG092C1VPN\n",
      "84 Ylr149c_EOG092C28\n",
      "85 TUB1_EOG092C2JRV\n",
      "86 NOP2_EOG091G0685\n",
      "87 ACS1_EOG092C0Z1N\n",
      "88 ECM13_EOG092C4UTV\n",
      "89 EFR3_EOG092C0OH2\n",
      "90 ENT3_EOG092C3ZA8\n",
      "91 MET4_EOG092C5FM7\n",
      "92 MGR3_EOG092C3NE2\n",
      "93 MSH6_EOG092C0EWT\n",
      "94 RKR1_EOG092C029X\n",
      "95 SXM1_EOG092C0CGO\n",
      "96 YHL026C_EOG092C2HBH\n",
      "97 AIM44_EOG092C5FAO\n",
      "98 DYN1_EOG092C00NZ\n",
      "99 EPO1_EOG092C0XF3\n",
      "100 KXD1_EOG092C5F1D\n",
      "101 MDV1_EOG092C172M\n",
      "102 MGA2_EOG092C06NJ\n",
      "103 NUM1_EOG092C01BH\n",
      "104 PBI1_EOG092C41LE\n",
      "105 SEC2_EOG092C3RPI\n",
      "106 SEC27_EOG092C10HM\n",
      "107 SEC63_EOG092C20BT\n",
      "108 SEC66_EOG092C58AK\n",
      "109 TOR1_EOG092C00HJ\n",
      "110 UBP1_EOG092C3TMU\n",
      "111 UBP2_EOG092C04TL\n",
      "112 VPS13_EOG092C0SXB\n",
      "113 YJL049W_EOG092C3Z1D\n",
      "114 YML003W_EOG092C00HJ\n",
      "115 YML002W_EOG092C087J\n",
      "116 YPT7_EOG092C5MDH\n",
      "117 YPT7_FUNGALx38\n",
      "118 EDE1muc_EOG092C055E\n",
      "119 Eps15_homologues391\n",
      "120 Opi1_blastx101\n",
      "121 Kri1_pfx2nr90_metazoa_x226\n",
      "122 Erb1_BOP1NT_NOT_EOG092C1381x904\n",
      "123 MCM3_28266497at2759_metazoa_nr50\n",
      "124 MCM3_28266497at2759_metazoa_nr90\n",
      "125 NOP2_plants_nr90\n",
      "126 Sec24C_nonfungi_nr90\n",
      "127 Sec24ABD_nonfungi_nr90\n",
      "128 TUBA_nonfungi_nr90\n",
      "129 Tcf25_metazoa_nr90\n",
      "130 Tcf25_viridiplantae_nr90\n",
      "131 Efr3_viridiplantae_nr90\n",
      "132 Efr3_metazoa_nr90\n",
      "133 Sxm1_all_nr90\n",
      "134 Msh6_Nterms\n",
      "135 Irr1_fungi_nr90\n",
      "136 Irr1_metazoa_nr90\n",
      "137 Vps13_Fungal_select_nr90\n",
      "138 CHMP7_all_nr90\n",
      "139 Kxd1_all_nr90\n",
      "140 Sec2_all_nr90\n",
      "141 bNOP2_nr50_n650_POG091H02E2\n",
      "142 KRI1_pfx2nr90_NOTmetazoa_x672\n",
      "143 USP36_all_nr90\n",
      "144 Vid27_domain_nonF_nr90\n",
      "145 Vid27_name_nonF_nr90\n",
      "146 Epsin_all_nr90\n",
      "147 Kri1_viridiplantae_nr90_x79\n",
      "148 Epsin_plants_nr90\n",
      "149 MSH6_nonFungi_nr50\n",
      " \n",
      "enter files to test (no commas) 148 149\n",
      " \n",
      "available analyses\n",
      "---------------\n",
      "0 Proteome Score Creates a csv file with a full list of FFAT scores for each protein\n",
      "1 Bottom Ten Creates a csv file the lowest 10 scores for each protein plus location\n",
      "2 Bottom Ten - Sorted Creates a csv file the lowest 10 scores for each protein, sorted by lowest score. Very slow\n",
      "3 Histogram Creates a csv file of a histogram of the FFAT scores of a proteome\n",
      "4 Histogram + randomised histogram (full proteome) a histogram of the FFAT scores of a proteome plus a randomised equivalent histogram (by full proteome)\n",
      "5 Histogram + randomised histogram (within protein) as 4, but randomised within protein\n",
      "6 Reordered Histogram a histogram of the FFAT scores of a proteome plus a reordered weighting equivalent histogram\n",
      "7 Histogram + randomised preserving acids (full proteome) score = 0.5 a histogram of the FFAT scores of a proteome plus a randomised equivalent preserving acids in low-scoring flanks score = 0.5\n",
      "8 Histogram + randomised preserving acids (within protein) score = 0.5 as 7, but randomised within protein\n",
      "9 Histogram + randomised preserving acids (full proteome) score = 0.0 a histogram of the FFAT scores of a proteome plus a randomised equivalent preserving acids in low-scoring flanks\n",
      "10 Histogram + randomised preserving acids (within protein) score = 0.0 as 7, but randomised within protein\n",
      "Enter analysis to test 5\n",
      " \n",
      "Enter 1 for full proteome file generation, anything else for just a histogram 99\n",
      "11: 20190214 19:29. running histogram_fully_randomised on proteome Epsin_plants_nr90\n",
      "12: 20190214 19:29. seed = 1\n",
      "13: 20190214 19:29. running histogram_fully_randomised on proteome MSH6_nonFungi_nr50\n",
      "14: 20190214 19:29. seed = 1\n",
      " \n",
      "finished analysis\n",
      " \n",
      "Enter 0 for live or 1 for test (anything else exits) 99\n",
      "terminated\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "****************************\n",
    "* master file for FFAT analysis of proteomes\n",
    "* simple arrangement\n",
    "* - select 'live' files - ie true proteomes, or test ones.\n",
    "* - select the particular file(s) for test\n",
    "* - select the particular analysis required.\n",
    "*\n",
    "* as analyses and files are added these are catered for in the modules called here\n",
    "* so I see little substantial change here\n",
    "* 15 Sep 2018\n",
    "*\n",
    "* here is the full (current structure)\n",
    "* lists of available files are kept in the following two files:\n",
    "* - proteomes.txt = 'live' versions\n",
    "* - proteomes_test_files.txt = test versions\n",
    "* Proteome Core file contains the class ProteomeReader for retrieving the files into a standard format\n",
    "*   and also contains ProteomeActivityLogger to keep track of what happened when\n",
    "* Proteome Scorer contains class SequenceAnalysis which does the basic work on the protein amino acid sequences\n",
    "* Proteome Analyser contains ProteomeAnalyser which does the high level analysis\n",
    "* this module allows with the user to specify what files are to be analysed, and retrieves the proteomes, \n",
    "*   and then allows the user to specify the analysis, It then passes the proteomes \n",
    "*   into ProteomeAnalyser for the specified analysis to take place\n",
    "*\n",
    "* Proteome Test Harness runs specific functional and full regression tests\n",
    "*\n",
    "* 29 Dec 2018\n",
    "* extra functionality both here and in Proteome Analyser\n",
    "* following request to add ability to add generation of proteome files and top ten for randomised proteomes\n",
    "****************************\n",
    "'''\n",
    "\n",
    "'''\n",
    "* - select the individual files to run - prod or test.\n",
    "* - normally should be core release versions\n",
    "'''\n",
    "\n",
    "#%run 'Proteome Core Release.ipynb'\n",
    "%run 'Proteome Core Dev.ipynb'\n",
    "%run 'Proteome Scorer Dev.ipynb'\n",
    "%run 'Proteome Analyser Dev.ipynb'\n",
    "\n",
    "def getnumberlist(strnos):\n",
    "    if '-' in strnos:\n",
    "        idx = strnos.index('-')\n",
    "        int1 = int(strnos[ : idx])\n",
    "        int2 = int(strnos[idx + 1:])\n",
    "        return range(int1, int2 + 1)\n",
    "    else:\n",
    "        return map(int, strnos.split())\n",
    "    \n",
    "def RunAnalysis():\n",
    "    \n",
    "    bexit = False\n",
    "    bprintmode = True\n",
    "    bhavetestlogger = False\n",
    "    bhaveprodlogger = False\n",
    "    \n",
    "    while not bexit:\n",
    "        \n",
    "        itestmode = int(raw_input('Enter 0 for live or 1 for test (anything else exits) '))\n",
    "\n",
    "        if itestmode == 0:\n",
    "            btestmode = False\n",
    "            if not bhaveprodlogger:\n",
    "                livelogger = ProteomeActivityLogger(btestmode, bprintmode)\n",
    "                bhaveprodlogger = True\n",
    "        elif itestmode == 1:\n",
    "            btestmode = True\n",
    "            if not bhavetestlogger:\n",
    "                testlogger = ProteomeActivityLogger(btestmode, bprintmode)\n",
    "                bhavetestlogger = True\n",
    "        else:\n",
    "            bexit = True\n",
    "\n",
    "        if not bexit:\n",
    "            \n",
    "            if btestmode:\n",
    "                mylogger = testlogger\n",
    "            else:\n",
    "                mylogger = livelogger\n",
    "\n",
    "            myreader = ProteomeReader(btestmode)\n",
    "            myscorer = SequenceAnalysis()\n",
    "            myanalysis = ProteomeAnalyser(mylogger, myscorer, btestmode)\n",
    "            \n",
    "            print 'reader version = ' + myreader.versionid\n",
    "            print 'scorer version = ' + myscorer.versionid\n",
    "            print 'analyser version = ' + myanalysis.versionid\n",
    "\n",
    "            mylogger.write_log_entry('core reader version = ' + myreader.versionid)\n",
    "\n",
    "            print 'available files'\n",
    "            print '---------------'\n",
    "            availablefiles = myreader.showlibrary()\n",
    "            for i in range(0, len(availablefiles)):\n",
    "                print i, availablefiles[i][0]\n",
    "\n",
    "            print ' '\n",
    "            fn = raw_input('enter files to test (no commas) ')\n",
    "            myfilenos = getnumberlist(fn)\n",
    "            #myfilenos =  map(int, fn.split())\n",
    "\n",
    "            availableanalyses = myanalysis.available_analysis()\n",
    "            print ' '\n",
    "            print 'available analyses'\n",
    "            print '---------------'\n",
    "            for i in range(0, len(availableanalyses)):\n",
    "                print i, availableanalyses[i][1], availableanalyses[i][2]\n",
    "\n",
    "            idx = int(raw_input('Enter analysis to test '))\n",
    "            print ' '\n",
    "            \n",
    "            bfullproteome = False # need as now an argument of run_analysis for some options\n",
    "            if myanalysis.has_generatedproteome_option(idx):\n",
    "                bfullproteome = (1 == int(raw_input('Enter 1 for full proteome file generation, anything else for just a histogram ')))\n",
    "\n",
    "            myproteomes = myreader.read_files(myfilenos)\n",
    "\n",
    "            myanalysis.run_analysis(idx, myproteomes, bfullproteome)\n",
    "            \n",
    "\n",
    "            print ' '\n",
    "            print 'finished analysis'\n",
    "            print ' '\n",
    "\n",
    "       \n",
    "    print 'terminated'\n",
    "        \n",
    "'''\n",
    "****************************\n",
    "* run RunAnalysis for a directed route to analysis\n",
    "****************************\n",
    "'''\n",
    "\n",
    "RunAnalysis()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
