{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\n****************************\\n* end of the analyser class\\n****************************\\n'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "****************************\n",
    "* Protein analyser class\n",
    "* contains the high level analysis routines applied to scoring and other low-level analysis done by \n",
    "* Proteome Scorer class\n",
    "* has a range of analyses.\n",
    "* descriptions are made available to the user, whoi selects an analysis to run\n",
    "* selected index is returned and appropriate analysis run\n",
    "* first version started 15 Sep 2018\n",
    "*\n",
    "* 22 Sep 2018 \n",
    "* added random seed setting capability, passed into an updated version of scorer\n",
    "* allows test mode and external control to be set and recorded\n",
    "* need to import time mopduile for this\n",
    "*\n",
    "* 4 Nov 2018\n",
    "* added ability to replace rogue amino acids\n",
    "* need arose due to an 'O' in Methanosarcina mazei pro\n",
    "*\n",
    "* 8 Nov 2018\n",
    "* changed the indexing of locations of the scores as the new version of the scorer class\n",
    "* this accessed by a variable bextended = True to add 'XXXXXX' on the front and 'XXXX' on the end\n",
    "* defaults to true\n",
    "*\n",
    "* 20 Nov 2018\n",
    "* wholesale changing of the histogram section\n",
    "* 22 Nov 2018 - take a cut - '1.0 Candidate\n",
    "****************************\n",
    "'''\n",
    "\n",
    "#import datetime - imported in Proteome Core\n",
    "#import random - this is imported in Proteome Scorer\n",
    "import csv\n",
    "import time\n",
    "\n",
    "class ProteomeAnalyser():\n",
    "    \n",
    "    def __init__(self, logger, scorer, btestmode = False):\n",
    "        \n",
    "        '''\n",
    "        ****************************\n",
    "        * initialises the Proteome analyser\n",
    "        * needs a logger to record what happens.\n",
    "        * and a scorer class to perform the analysis\n",
    "        * hence cannot be initiatied without either of these.\n",
    "        * note the logger comes from Proteome core so we can assume functions datestamp() and timestamp() exist\n",
    "        ****************************\n",
    "        '''\n",
    "        \n",
    "        self.logger = logger\n",
    "        self.scorer = scorer\n",
    "        self.versionid = '1.0'\n",
    "        self.filepath = '/Users/johnslee/Documents/'\n",
    "        \n",
    "        if btestmode:\n",
    "            self.randomseed = 1\n",
    "        else:\n",
    "            self.randomseed = int(time.time())\n",
    "        \n",
    "        versionlist = 'scorer version = ' + scorer.versionid\n",
    "        versionlist += ', logger version = ' + self.versionid\n",
    "        self.logger.write_log_entry(versionlist)\n",
    "        self.logger.write_log_entry('random seed set to ' + str(self.randomseed))\n",
    "        \n",
    "        self.numberofbuckets = 37# module-variable for the number of buckets in a histogram\n",
    "        \n",
    "        \n",
    "    def available_analysis(self):\n",
    "        \n",
    "        '''\n",
    "        ****************************\n",
    "        * returns a list of available analysis\n",
    "        * each number must be matched by an entry in function run_analysis()\n",
    "        * which is at the bottom of the class\n",
    "        ****************************\n",
    "        '''\n",
    "        \n",
    "        myanalyses = []\n",
    "        myanalyses.append([0, 'Proteome Score','Creates a csv file with a full list of FFAT scores for each protein'])\n",
    "        myanalyses.append([1, 'Bottom Ten','Creates a csv file the lowest 10 scores for each protein plus location'])\n",
    "        myanalyses.append([2, 'Bottom Ten - Sorted','Creates a csv file the lowest 10 scores for each protein, sorted by lowest score. Very slow'])\n",
    "        myanalyses.append([3, 'Histogram','Creates a csv file of a histogram of the FFAT scores of a proteome'])\n",
    "        myanalyses.append([4, 'Histogram + randomised histogram (full proteome)','a histogram of the FFAT scores of a proteome plus a randomised equivalent histogram (by full proteome)'])\n",
    "        myanalyses.append([5, 'Histogram + randomised histogram (by protein)','a histogram of the FFAT scores of a proteome plus a randomised equivalent histogram (by protein)'])\n",
    "        myanalyses.append([6, 'Reordered Histogram','a histogram of the FFAT scores of a proteome plus a reordered weighting equivalent histogram'])\n",
    "        myanalyses.append([7, 'Histogram + rendomised preserving acids','a histogram of the FFAT scores of a proteome plus a randomised equivalent preserving acids in low-scoring flanks' ])\n",
    "        return myanalyses\n",
    "        \n",
    "        \n",
    "    def create_proteomescore_csv(self, proteomes):\n",
    "        \n",
    "        '''\n",
    "        ****************************\n",
    "        * takes a list of proteomes\n",
    "        * for each proteome\n",
    "        * scores the protein and makes a csv file from it.\n",
    "        *\n",
    "        * analysis details:\n",
    "        * number - 0\n",
    "        * title -  'Proteome Score'\n",
    "        * description - 'Creates a csv file with a full list of FFAT scores for each protein'\n",
    "        *\n",
    "        ****************************\n",
    "        '''\n",
    "        \n",
    "        bextended = True\n",
    "        \n",
    "        for proteome in proteomes:\n",
    "        \n",
    "            orgfilename = proteome[0][0]\n",
    "            \n",
    "            self.logger.write_log_entry('running proteome score on proteome ' + orgfilename)\n",
    "    \n",
    "            foutname = self.filepath + orgfilename + '_rawscores_' + datestamp() + '.csv'\n",
    "            fout = open(foutname,\"w\")\n",
    "            cr = csv.writer(fout, delimiter=',',quoting=csv.QUOTE_ALL)\n",
    "           \n",
    "            myproteins = proteome[1]\n",
    "    \n",
    "            ctr = 0\n",
    "\n",
    "            for protein in myproteins:\n",
    "\n",
    "                ctr += 1\n",
    "\n",
    "                myresults = self.scorer.score_protein(protein[1], bextended)\n",
    "                mycsv = [protein[0][0],protein[0][1]]\n",
    "                for result in myresults:\n",
    "                    mycsv.append(result)\n",
    "                cr.writerow(mycsv)\n",
    "            \n",
    "            self.logger.write_log_entry('finished scoring ' + orgfilename + ' - ' + str(ctr) + ' proteins')\n",
    "            if bextended:\n",
    "                self.logger.write_log_entry('scoring done with X amino acids appended front and back')\n",
    "            \n",
    "            fout.close()\n",
    "            \n",
    "        self.logger.write_log_entry('finished proteome score')\n",
    "        \n",
    "    def write_generated_proteome_file(self, proteins, filename):\n",
    "        '''\n",
    "        ********************************\n",
    "        * 22 Sep 2018\n",
    "        * creates a txt file of a generated file in a Fasta like way\n",
    "        * key is to avoid confusing this with an actual one\n",
    "        * and to write in a way that can easily be read.\n",
    "        ********************************\n",
    "        '''\n",
    "        \n",
    "        self.logger.write_log_entry('writing randomised proteome ' + filename)\n",
    "    \n",
    "        foutname = self.filepath + filename + datestamp() + '.txt'\n",
    "        fout = open(foutname,\"w\")\n",
    "    \n",
    "        for protein in proteins:\n",
    "\n",
    "            # construct Fasta-like header\n",
    "            # >YAL010C MDM10 SGDID:S000000008, Chr I fr...\n",
    "\n",
    "            proteinheader = '>' + protein[0][0] + ' ' + protein[0][1] + ' randomised version'\n",
    "            fout.write(proteinheader + '\\n')\n",
    "            fout.write(protein[1] + '\\n')\n",
    "            \n",
    "        fout.close()\n",
    "        \n",
    "    def bottom_ten(self, proteomes, bextended = True):\n",
    "\n",
    "        '''\n",
    "        ********************************\n",
    "        * scores each protein, and produces a list of the bottome 10 scores plus starting location\n",
    "        * doesn't sort - much faster than sorting, which can be done in excel very easily.\n",
    "        *\n",
    "        * analysis details:\n",
    "        * number - 1\n",
    "        * title -  'Bottom Ten'\n",
    "        * description - 'Creates a csv file the lowest 10 scores for each protein plus location'\n",
    "        *\n",
    "        * 17 Sep added limit to protein_score_ordered to speed up the function\n",
    "        *\n",
    "        * 8 Nov 18 added bextension to indicate 'X's added front and backto allow scoring of start and end sequences\n",
    "        ********************************\n",
    "        '''\n",
    "\n",
    "        numscores = 10\n",
    "        \n",
    "        # now shift index for the start of the sequence depending whether been extended with 'X 8 Nov 18\n",
    "        if bextended:\n",
    "            istartindex = 6\n",
    "        else:\n",
    "            istartindex = 0\n",
    "\n",
    "        for proteome in proteomes:\n",
    "\n",
    "            orgfilename = proteome[0][0]\n",
    "\n",
    "            foutname = self.filepath + orgfilename + '_bottomten_' + datestamp() + '.csv'\n",
    "            fout = open(foutname,\"w\")\n",
    "            cr = csv.writer(fout, delimiter=',',quoting=csv.QUOTE_ALL)\n",
    "            \n",
    "            self.logger.write_log_entry('running bottom_ten on proteome ' + orgfilename)\n",
    "\n",
    "            myproteins = proteome[1]\n",
    "            allmybottomten = []\n",
    "\n",
    "            for protein in myproteins:\n",
    "                \n",
    "                myresults = self.scorer.score_protein(protein[1], bextended)\n",
    "                myrankedresults = self.scorer.protein_score_ordered(myresults, numscores)\n",
    "                thisbottomten = myrankedresults[0 : numscores]\n",
    "                \n",
    "                mycsv = []\n",
    "                mycsv.append(protein[0][0])\n",
    "                mycsv.append(protein[0][1])\n",
    "                mycsv.append(len(protein[1]))\n",
    "                \n",
    "                for i in range(0, numscores):\n",
    "                    if i < len(thisbottomten):\n",
    "                        mycsv.append(thisbottomten[i][0])\n",
    "                    \n",
    "                mycsv.append('')\n",
    "                \n",
    "                for i in range(0, numscores):\n",
    "                    if i < len(thisbottomten):\n",
    "                        mycsv.append(thisbottomten[i][1] - istartindex)\n",
    "                \n",
    "                cr.writerow(mycsv)\n",
    "                \n",
    "            self.logger.write_log_entry('finished sorted and ranking proteins in ' + orgfilename)\n",
    "                \n",
    "            fout.close()\n",
    "            \n",
    "        self.logger.write_log_entry('finished bottom ten')\n",
    "        \n",
    "    def bottom_ten_sorted(self, proteomes, bextended = True):\n",
    "\n",
    "        '''\n",
    "        ********************************\n",
    "        * scores each protein, and produces a list of the bottome 10 scores plus starting location\n",
    "        * then sorts on first score and prints lowest scores first\n",
    "        * a future adaptation would be to sort on the first ten starting in positon 10\n",
    "        * so the first one is the lowest of all.\n",
    "        *\n",
    "        * analysis details:\n",
    "        * number - 2\n",
    "        * title -  'Bottom Ten - sorted'\n",
    "        * description - 'Creates a csv file the lowest 10 scores for each protein, sorted by lowest score'\n",
    "        *\n",
    "        * 8 Nov 18 added bextension to indicate 'X's added front and backto allow scoring of start and end sequences\n",
    "        *\n",
    "        ********************************\n",
    "        '''\n",
    "\n",
    "        numscores = 10\n",
    "        \n",
    "        # now shift index for the start of the sequence depending whether been extended with 'X 8 Nov 18\n",
    "        if bextended:\n",
    "            istartindex = 6\n",
    "        else:\n",
    "            istartindex = 0\n",
    "\n",
    "        for proteome in proteomes:\n",
    "\n",
    "            orgfilename = proteome[0][0]\n",
    "\n",
    "            foutname = self.filepath + orgfilename + '_bottomten_sorted_' + datestamp() + '.csv'\n",
    "            fout = open(foutname,\"w\")\n",
    "            cr = csv.writer(fout, delimiter=',',quoting=csv.QUOTE_ALL)\n",
    "            \n",
    "            self.logger.write_log_entry('running bottom_ten_sorted on proteome ' + orgfilename)\n",
    "\n",
    "            myproteins = proteome[1]\n",
    "            allmybottomten = []\n",
    "\n",
    "            for protein in myproteins:\n",
    "                \n",
    "                myresults = self.scorer.score_protein(protein[1], bextended)\n",
    "                myrankedresults = self.scorer.protein_score_ordered(myresults, numscores)\n",
    "                thisbottomten = myrankedresults[0 : numscores]\n",
    "                allmybottomten.append(thisbottomten)\n",
    "                \n",
    "            # now have all the bottom 10 scores. need to sort\n",
    "            \n",
    "            numproteins = len(allmybottomten)\n",
    "            seq = range(0, numproteins)\n",
    "            \n",
    "            #for irank in range(numscores - 1, -1, -1):\n",
    "            for irank in [1,0]:# doing all 10 very time consuming\n",
    "                # irank is the number in the list we are sorting on.\n",
    "                # we sort by 0th, then 1st, etc, so we need to sort in reverse order\n",
    "                \n",
    "                for i in range(0, numproteins - 1):\n",
    "                    for j in range(numproteins - 2, i - 1, -1):\n",
    "                        if allmybottomten[seq[j]][irank][0] > allmybottomten[seq[j + 1]][irank][0]:\n",
    "                            temp = seq [j]\n",
    "                            seq[j] = seq[j + 1]\n",
    "                            seq[j + 1] = temp\n",
    "                \n",
    "            # now have the sequence for this protein\n",
    "                \n",
    "            for s in seq:\n",
    "                \n",
    "                mycsv = []\n",
    "                mycsv.append(myproteins[s][0][0])\n",
    "                mycsv.append(myproteins[s][0][1])\n",
    "                mycsv.append(len(myproteins[s][1]))\n",
    "                mycsv.append('')\n",
    "                \n",
    "                for i in range(0, numscores):\n",
    "                    if i < len(allmybottomten[s]):\n",
    "                        mycsv.append(allmybottomten[s][i][0])\n",
    "                    \n",
    "                mycsv.append('')\n",
    "                \n",
    "                for i in range(0, numscores):\n",
    "                    if i < len(allmybottomten[s]):\n",
    "                        mycsv.append(allmybottomten[s][i][1] - istartindex)                           \n",
    "                \n",
    "                cr.writerow(mycsv)\n",
    "                \n",
    "            self.logger.write_log_entry('finished sorted and ranking proteins in ' + orgfilename)\n",
    "                \n",
    "            fout.close()\n",
    "            \n",
    "        self.logger.write_log_entry('finished bottom ten - sorted')\n",
    "        \n",
    "    '''\n",
    "    ********************************\n",
    "    * now a set of routines round histograms\n",
    "    * this counts the occurence of FFAT scores for a set of proteins\n",
    "    *\n",
    "    * basis is make_histogram - takes a proteome and generates a list structure with a histogram count.\n",
    "    ********************************\n",
    "    '''\n",
    "    \n",
    "    def make_histogram(self, proteome):\n",
    "        \n",
    "        '''\n",
    "        ********************************\n",
    "        * creates a histogram of FFAT scores for a proteome\n",
    "        * 20 Nov 2018\n",
    "        * generates a list of scores and count.\n",
    "        * generally first is 0, and there are sequences that score 0\n",
    "        * then 0.5, and anything between 0 and 0.5 gets into that bucket \n",
    "        * does a single proteome\n",
    "        * and returns as a list of [proteome_detail,[[score, count], [score, count]]]\n",
    "        ********************************\n",
    "        '''\n",
    "        \n",
    "        bars = []\n",
    "        for i in range(0,self.numberofbuckets):\n",
    "            bars.append([i/2.0, 0])\n",
    "\n",
    "        orgdetails = proteome[0]\n",
    "\n",
    "        myproteins = proteome[1]\n",
    "\n",
    "        for protein in myproteins:\n",
    "\n",
    "            myresults = self.scorer.score_protein(protein[1])\n",
    "\n",
    "            for res in myresults:\n",
    "                idx = int(res * 2.0)\n",
    "                if idx < len(bars):\n",
    "                    bars[idx][1] += 1\n",
    "                else:\n",
    "                    print 'score beyond high bound ' + str(res) \n",
    "\n",
    "        return [orgdetails,bars]\n",
    "    \n",
    "    def proteome_histogram(self, proteomes):\n",
    "        \n",
    "        for proteome in proteomes:\n",
    "            \n",
    "            buckets = self.make_histogram(proteome)\n",
    "            \n",
    "            orgfilename = proteome[0][0]\n",
    "            \n",
    "            foutname = self.filepath + orgfilename + '_histogram_' + datestamp() + '.csv'\n",
    "            fout = open(foutname,\"w\")\n",
    "            cr = csv.writer(fout, delimiter=',',quoting=csv.QUOTE_ALL)\n",
    "\n",
    "            for bucket in buckets[1]:\n",
    "                cr.writerow(bucket)\n",
    "        \n",
    "    \n",
    "    def histogram_randomised(self, proteomes, originalseed, numhistograms, bbyprotein = False):\n",
    "        \n",
    "        '''\n",
    "        ********************************\n",
    "        * creates a histogram of FFAT scores for a proteome\n",
    "        * writes to a .txt file\n",
    "        * compares to a randomised version\n",
    "        * includes amino acids count check\n",
    "        *\n",
    "        * 13 Nov 18\n",
    "        * doing a single randomisation not sufficient - need to do a number\n",
    "        * also include a seed, which will be incremented by 1 each loop\n",
    "        * change in output to an array histograms[[bars][numbers in bar]]\n",
    "        ********************************\n",
    "        '''\n",
    "        \n",
    "        csvtoprow = ['bucket','count']\n",
    "        for i in range(0,numhistograms):\n",
    "            csvtoprow.append('sim #' + str(i))\n",
    "\n",
    "        for proteome in proteomes:\n",
    "\n",
    "            orgdetails = proteome[0]\n",
    "            orgfilename = orgdetails[0]\n",
    "            myproteins = proteome[1]\n",
    "            \n",
    "            csvtitle = ['histogram_randomised','seed = ' + str(originalseed), orgfilename, datestamp()]\n",
    "            self.logger.write_log_entry('running histogram_randomised on proteome ' + orgfilename)\n",
    "            \n",
    "            if bbyprotein:\n",
    "                foutname = self.filepath + orgfilename + '_histogram_by_protein_' + datestamp() + '.csv'\n",
    "            else:\n",
    "                foutname = self.filepath + orgfilename + '_histogram_full_proteome_' + datestamp() + '.csv'\n",
    "                \n",
    "            fout = open(foutname,\"w\")\n",
    "            cr = csv.writer(fout, delimiter=',',quoting=csv.QUOTE_ALL)\n",
    "            \n",
    "            originalbars = self.make_histogram(proteome)\n",
    "        \n",
    "            lastscore = 0 # make element 2 a running count.\n",
    "            \n",
    "            randombars = []\n",
    "            \n",
    "            currentseed = originalseed\n",
    "            \n",
    "            for j in range(0, numhistograms):\n",
    "                \n",
    "                myrandomisedproteins = self.scorer.randomise_proteins(myproteins, currentseed, bbyprotein)\n",
    "                \n",
    "                myrandomisedproteome = [orgdetails, myrandomisedproteins]\n",
    "                \n",
    "                randombars.append(self.make_histogram(myrandomisedproteome))\n",
    "                            \n",
    "                currentseed += 1\n",
    "            \n",
    "            cr.writerow(csvtitle)\n",
    "            cr.writerow(csvtoprow)\n",
    "            \n",
    "            for i in range(0, self.numberofbuckets):\n",
    "                \n",
    "                mycsv = []\n",
    "                mycsv.append(originalbars[1][i][0])\n",
    "                mycsv.append(originalbars[1][i][1])\n",
    "                for j in range(0, numhistograms):\n",
    "                    mycsv.append(randombars[j][1][i][1])\n",
    "               \n",
    "                cr.writerow(mycsv)\n",
    "            \n",
    "    def histogram_reordered(self, proteomes, reorder):\n",
    "        \n",
    "        '''\n",
    "        ********************************\n",
    "        * creates a histogram of FFAT scores for a proteome\n",
    "        * writes to a .txt file\n",
    "        * reorderers the scoring of the main sequence according to the variable reorder\n",
    "        * compares to the original version\n",
    "        * includes amino acids count check\n",
    "        ********************************\n",
    "        '''\n",
    "        \n",
    "        # check the reorder meets the requirements of reordering\n",
    "        # each number from 0 to 6 once and only once.\n",
    "        \n",
    "        standard =range(0,7)\n",
    "        if sorted(reorder) <> standard:\n",
    "            print 'warning - reorder sequence does not meet requirements of each number between 0 and 6 once and only once'\n",
    "        else:\n",
    "            \n",
    "            seq = ''\n",
    "            for r in reorder:\n",
    "                seq += str(r + 1)\n",
    "                \n",
    "            bars = []\n",
    "            numbars = 37\n",
    "    \n",
    "            for i in range(0,numbars):\n",
    "                bars.append([i/2.0,0,0])\n",
    "    \n",
    "            for proteome in proteomes:\n",
    "\n",
    "                orgfilename = proteome[0][0]\n",
    "\n",
    "                foutname = self.filepath + orgfilename + '_histogram_reordered_' + seq + '_' + datestamp() + '.txt'\n",
    "                fout = open(foutname,\"w\")\n",
    "\n",
    "                self.logger.write_log_entry('running histogram plus reordered histogram - results in ' + foutname)\n",
    "\n",
    "                myproteins = proteome[1]\n",
    "\n",
    "                ctrscore = 0\n",
    "\n",
    "                for i in range(0,numbars):\n",
    "                    bars[i][1] = 0\n",
    "\n",
    "                for protein in myproteins:\n",
    "\n",
    "                    myresults = self.scorer.score_protein(protein[1])\n",
    "                    ctrscore += len(myresults)\n",
    "                    for res in myresults:\n",
    "                        idx = int(res * 2)\n",
    "                        if idx < len(bars):\n",
    "                            bars[idx][1] += 1\n",
    "                        else:\n",
    "                            print('score beyond high bound ' + str(res))\n",
    "\n",
    "                fout.write(orgfilename + ' histogram - ' + str(ctrscore) + ' scores' + '\\n')\n",
    "                fout.write('-------------\\n')\n",
    "\n",
    "                lastscore = 0 # make element 2 a running count.\n",
    "                for bar in bars:\n",
    "                    bar[2] = lastscore + 100.0 * bar[1]/ctrscore\n",
    "                    lastscore = bar[2]\n",
    "                    fout.write(str(bar[0]) + ': ' + str(bar[1]) + ', ' + str(bar[2])[:6] + '%\\n')\n",
    "                fout.write('-------------\\n\\n')\n",
    "\n",
    "                for i in range(0,numbars):\n",
    "                    bars[i][1] = 0\n",
    "                    \n",
    "                #use ctrscore as the total of all scores again as shouldn't change just on re-ordering.\n",
    "\n",
    "                for protein in myproteins:\n",
    "\n",
    "                    myresults = self.scorer.score_protein_reordered(protein[1], reorder)\n",
    "                    for res in myresults:\n",
    "                        idx = int(res * 2)\n",
    "                        if idx < len(bars):\n",
    "                            bars[idx][1] += 1\n",
    "                        else:\n",
    "                            print('score beyond high bound ' + str(res))\n",
    "\n",
    "                fout.write(orgfilename + ' histogram - reordered - ' + seq + '\\n')\n",
    "                fout.write('-------------\\n')\n",
    "\n",
    "                lastscore = 0 # make element 2 a running count.\n",
    "                for bar in bars:\n",
    "                    bar[2] = lastscore + 100.0 * bar[1]/ctrscore\n",
    "                    lastscore = bar[2]\n",
    "                    fout.write(str(bar[0]) + ': ' + str(bar[1]) + ', ' + str(bar[2])[:6] + '%\\n')\n",
    "                fout.write('-------------\\n')\n",
    "\n",
    "            self.logger.write_log_entry('finished running histogram routine ')\n",
    "            self.logger.write_log_entry('-----------------------------------')\n",
    "             \n",
    "    def histogram_randomised_preserved_acidflank(self, proteomes,originalseed, numhistograms, cutoff = 0.5, bbyprotein = False):\n",
    "        \n",
    "        '''\n",
    "        ********************************\n",
    "        * creates a histogram of FFAT scores for a proteome\n",
    "        * writes to a .txt file\n",
    "        * reorders apart from preserving qualifying acids\n",
    "        * which are acids 'DEST' in the flank of segments with FFAT scores of 0.5 or less\n",
    "        * compares to the original version\n",
    "        * includes amino acids count check\n",
    "        *\n",
    "        * 22 Sep 2018 - make a separate file of the hits\n",
    "        *\n",
    "        * 20 Nov 18 - make it jus a flank score with either 0 or 0.5 \n",
    "        * introduce the cutoff score as an option.\n",
    "        ********************************\n",
    "        '''\n",
    "        csvtoprow = ['bucket','count']\n",
    "        for i in range(0,numhistograms):\n",
    "            csvtoprow.append('sim #' + str(i))\n",
    "\n",
    "        for proteome in proteomes:\n",
    "\n",
    "            orgdetails = proteome[0]\n",
    "            orgfilename = orgdetails[0]\n",
    "            myproteins = proteome[1]\n",
    "            \n",
    "            csvtitle = ['histogram_randomised_preserved_acidflank','seed = ' + str(originalseed), orgfilename,'cut off = '+ str(cutoff), datestamp()]\n",
    "            self.logger.write_log_entry('running histogram_randomised_preserved_acidflank on proteome ' + orgfilename)\n",
    "            self.logger.write_log_entry('seed = ' + str(originalseed) + ',  cut off = '+ str(cutoff))\n",
    "            \n",
    "            if bbyprotein:\n",
    "                foutname = self.filepath + orgfilename + '_acidflankmask_histogram_by_protein_' + datestamp() + '.csv'\n",
    "            else:\n",
    "                foutname = self.filepath + orgfilename + '_acidflankmask_histogram_full_proteome_' + datestamp() + '.csv'\n",
    "                \n",
    "            fout = open(foutname,\"w\")\n",
    "            cr = csv.writer(fout, delimiter=',',quoting=csv.QUOTE_ALL)\n",
    "            \n",
    "            originalbars = self.make_histogram(proteome)\n",
    "        \n",
    "            lastscore = 0 # make element 2 a running count.\n",
    "            \n",
    "            randombars = []\n",
    "            \n",
    "            currentseed = originalseed\n",
    "            \n",
    "            for j in range(0, numhistograms):\n",
    "                \n",
    "                myrandomisedproteins = self.scorer.randomise_proteins_preserve_acidic_flank(myproteins, currentseed, cutoff, bbyprotein)\n",
    "                \n",
    "                myrandomisedproteome = [orgdetails, myrandomisedproteins]\n",
    "                \n",
    "                randombars.append(self.make_histogram(myrandomisedproteome))\n",
    "                            \n",
    "                currentseed += 1\n",
    "            \n",
    "            cr.writerow(csvtitle)\n",
    "            cr.writerow(csvtoprow)\n",
    "            \n",
    "            for i in range(0, self.numberofbuckets):\n",
    "                \n",
    "                mycsv = []\n",
    "                mycsv.append(originalbars[1][i][0])\n",
    "                mycsv.append(originalbars[1][i][1])\n",
    "                for j in range(0, numhistograms):\n",
    "                    mycsv.append(randombars[j][1][i][1])\n",
    "               \n",
    "                cr.writerow(mycsv)\n",
    "                \n",
    "        \n",
    "    def run_analysis(self, analysisnumber, proteomes):\n",
    "        \n",
    "        if analysisnumber == 0:\n",
    "            self.create_proteomescore_csv(proteomes)\n",
    "        elif analysisnumber == 1:\n",
    "            self.bottom_ten(proteomes)\n",
    "        elif analysisnumber == 2:\n",
    "            self.bottom_ten_sorted(proteomes)\n",
    "        elif analysisnumber == 3:\n",
    "            self.proteome_histogram(proteomes)\n",
    "        elif analysisnumber == 4:\n",
    "            numhistograms = 10\n",
    "            thisseed = 1\n",
    "            self.histogram_randomised(proteomes, thisseed, numhistograms, False )\n",
    "        elif analysisnumber == 5:\n",
    "            numhistograms = 10\n",
    "            thisseed = 1\n",
    "            self.histogram_randomised(proteomes, thisseed, numhistograms, True)\n",
    "        elif analysisnumber == 6:\n",
    "            newsequence = [2,5,6,1,0,4,3]\n",
    "            self.histogram_reordered(proteomes, newsequence)\n",
    "        elif analysisnumber == 7:\n",
    "            numhistograms = 10\n",
    "            thisseed = 1\n",
    "            #histogram_randomised_preserved_acidflank(self, proteomes, originalseed, numhistograms, cutoff = 0.5, bbyprotein = False)\n",
    "            self.histogram_randomised_preserved_acidflank(proteomes, thisseed, numhistograms, 0.5, False)\n",
    "        else:\n",
    "            print 'unrecognised analysis index '+ str(analysisnumber)\n",
    "        \n",
    "        \n",
    "'''\n",
    "****************************\n",
    "* end of the analyser class\n",
    "****************************\n",
    "'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
