{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Core GH 22 03\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "****************************\n",
    "* Proteome Core\n",
    "* \n",
    "* main access is through function readfiles() which takes a list of integers\n",
    "* the function then returns a list of file contents with each element being that of the corresponding item in the list\n",
    "* the numbers refer to the sequence of files in a text file currently living at /Users/johnslee/proteomes.txt\n",
    "* for each item in the list readfiles returns a two element list\n",
    "* [text element from the list of files][proteome1, proteome2, proteome3,]]\n",
    "* each proteome consists of a list of proteins[protein1, protein2, protein3]\n",
    "* each protein consists of a header and the code\n",
    "* header = [protein name, protein reference]\n",
    "*\n",
    "****************************\n",
    "'''\n",
    "import datetime\n",
    "\n",
    "def datestamp():\n",
    "    return datetime.date.today().strftime(\"%Y%m%d\")\n",
    "\n",
    "def timestamp():\n",
    "    return datetime.datetime.now().strftime(\"%H:%M\")\n",
    "\n",
    "def getnumberlist(strnos):\n",
    "    if '-' in strnos:\n",
    "        idx = strnos.index('-')\n",
    "        int1 = int(strnos[ : idx])\n",
    "        int2 = int(strnos[idx + 1:])\n",
    "        return range(int1, int2 + 1)\n",
    "    else:\n",
    "        return map(int, strnos.split())\n",
    "\n",
    "class ProteomeActivityLogger():\n",
    "    \n",
    "    '''\n",
    "    ****************************\n",
    "    * logging class that allows other programmes to create a log of activity.\n",
    "    * has two modes - test and prod - ie non test\n",
    "    * creates a new log each day, or opens the existing log for that day.\n",
    "    * has a printmode that allows all entries going to the log file to be printd on screen\n",
    "    ****************************\n",
    "    '''\n",
    "    \n",
    "    def __init__(self, testmode = False, printmode = False):\n",
    "    \n",
    "        self.versionid = '22.0 GH '\n",
    "        self.fpath = '/Users/johnslee/Documents/'\n",
    "        self.counter = 0\n",
    "        self.printmode = printmode\n",
    "    \n",
    "        if testmode:\n",
    "            fname = self.fpath + 'ProteomeTestlog' + datestamp() + '.txt'\n",
    "            firstline = 'New test log'\n",
    "        else:\n",
    "            fname = self.fpath + 'Proteomelog' + datestamp() + '.txt'\n",
    "            firstline = 'New activity log'\n",
    "        \n",
    "        self.f = open(fname, 'a')\n",
    "        \n",
    "        firstline += ' version ' + self.versionid\n",
    "        self.f.write(firstline + '\\n')  # python will convert \\n to os.linesep \n",
    "        if self.printmode:\n",
    "            print (firstline)\n",
    "    \n",
    "    def write_log_entry(self, entry):\n",
    "        \n",
    "        premlim = str(self.counter) + ': ' + datestamp() + ' ' + timestamp() + '. '\n",
    "        self.f.write(premlim + entry + '\\n')\n",
    "        self.counter += 1\n",
    "        if self.printmode:\n",
    "            print (premlim + entry)\n",
    "        \n",
    "class ProteomeReader():\n",
    "    \n",
    "    '''\n",
    "    ****************************\n",
    "    * class ProteomeReader\n",
    "    * reads proteome files \n",
    "    * maintains a list of files taken from a text file proteomes.txt\n",
    "    * so new proteomes are added in that file.\n",
    "    * info in file is [filename,organism name,readindex]\n",
    "    * has four type of files it handles.\n",
    "    * readindex = 0 for fasta, 1 = uniref, 2 = dashed, 3 = OrthoDB, 4 = ACBD5\n",
    "    ****************************\n",
    "    '''\n",
    "    \n",
    "    def __init__(self):\n",
    "        \n",
    "        '''\n",
    "        ***************************\n",
    "        * builds a library of available proteomes\n",
    "        * assumes all have the same path and type\n",
    "        * also sets the version control whgich should be changed by the programmer for each release.\n",
    "        * contains the file path in a variable set here.\n",
    "        ***************************\n",
    "        '''\n",
    "        \n",
    "        self.versionid = '22.0 GH'\n",
    "        self.fpath = '/Users/johnslee/Documents/'\n",
    "        self.ftype = '.txt'\n",
    "        \n",
    "    '''\n",
    "    ****************************\n",
    "    * 2 Sep 2019\n",
    "    * some new routines to move file library, file selecion and readin into the module\n",
    "    *\n",
    "    * calling process\n",
    "    *   myreader = ProteomeReader()\n",
    "    *   bexit = myreader.select_file_library()   \n",
    "    *   if not bexit:\n",
    "    *       myproteomes = myreader.select_files()\n",
    "    *\n",
    "    * and calling function now has the full proteomes\n",
    "    ****************************\n",
    "    '''\n",
    "        \n",
    "    def select_file_library(self, bprintmode = True):\n",
    "        \n",
    "        '''\n",
    "        ****************************\n",
    "        * allows the user to select a file library\n",
    "        * also the means of exiting\n",
    "        * if user selects something outside the offered range, returns True\n",
    "        * otherwise returns (True if 0 files)\n",
    "        ****************************\n",
    "        '''\n",
    "        \n",
    "        print ('File selection:')\n",
    "        print ('0 for Main file list ')\n",
    "        print ('1 for Test files')\n",
    "        print (' ')\n",
    "        \n",
    "        ifilecategory = int(input('Anything else exits '))\n",
    "\n",
    "        if ifilecategory == 0:\n",
    "       \n",
    "            filenameproteomes = '/Users/johnslee/proteomes.txt'\n",
    "            self.activelogger= ProteomeActivityLogger(False, bprintmode)\n",
    "       \n",
    "        elif ifilecategory == 1:\n",
    "            \n",
    "            filenameproteomes = '/Users/johnslee/proteomes_test_files.txt'\n",
    "            self.activelogger= ProteomeActivityLogger(True, bprintmode)\n",
    "            \n",
    "        else:\n",
    "            \n",
    "            # exiting\n",
    "            return True\n",
    "        \n",
    "        self.proteomelibrary = []\n",
    "        \n",
    "        f = open(filenameproteomes, 'r')\n",
    "        \n",
    "        for line in f:\n",
    "            \n",
    "            thisfile = line[:]\n",
    "            \n",
    "            if ',' in thisfile:\n",
    "                idx = thisfile.index(',')\n",
    "                thisfilename = thisfile[ : idx]\n",
    "                \n",
    "            thisfile = thisfile[ idx + 1 :]\n",
    "            \n",
    "            if ',' in thisfile:\n",
    "                idx = thisfile.index(',')\n",
    "                thisfiledescription = thisfile[ : idx]\n",
    "                \n",
    "            thisfile = thisfile[ idx + 1:].strip()\n",
    "\n",
    "            thisfilemode = int(thisfile)\n",
    "            \n",
    "            self.proteomelibrary.append([thisfilename,thisfiledescription,thisfilemode])\n",
    "        \n",
    "        return len(self.proteomelibrary) == 0\n",
    "            \n",
    "    def select_files(self):\n",
    "        \n",
    "        print (' ')\n",
    "        print ('available files')\n",
    "        print ('---------------')\n",
    "        \n",
    "        availablefiles = self.proteomelibrary\n",
    "        \n",
    "        for i in range(0, len(availablefiles)):\n",
    "            print (i, availablefiles[i][0])\n",
    "\n",
    "        print (' ')\n",
    "        fn = input('enter files to test (no commas) ') # changed from raw_input - python 3 upgrade\n",
    "        \n",
    "        myfilenos = getnumberlist(fn)\n",
    "\n",
    "        myproteomes = self.read_files(myfilenos)\n",
    "        \n",
    "        return myproteomes\n",
    "        \n",
    "    def read_files(self, flist):\n",
    "        '''\n",
    "        ****************************\n",
    "        * 28 Aug 2018\n",
    "        * rejig of how protein files are read in\n",
    "        * flist is a list of numbers referring to indices in the file library\n",
    "        * reads in the content and returns a single list\n",
    "        * with one entry per file\n",
    "        * so returns [file1, file2]\n",
    "        * file 1 = [proteome library 1, proteins]\n",
    "        * proteome library = ['file name','organism name',read method]\n",
    "        * proteins = [protein1, protein2]\n",
    "        * protein1 = [[protein ID, protein name, comments],[code]]\n",
    "        * so the code for the 5th protein of the 3rd file is in proteins[2][1][4][1]\n",
    "        ****************************\n",
    "        '''\n",
    "    \n",
    "        proteins = []\n",
    "        for f in flist:\n",
    "            myfile = self.proteomelibrary[f]\n",
    "            fullfilename = self.fpath + myfile[0] + self.ftype\n",
    "            if myfile[2] == 0:\n",
    "                myproteins = self.readFASTA(fullfilename)\n",
    "            elif myfile[2] == 1:\n",
    "                myproteins = self.readUniRef50(fullfilename)\n",
    "            elif myfile[2] == 2:\n",
    "                myproteins = self.readDashed(fullfilename)\n",
    "            elif myfile[2] == 3:\n",
    "                myproteins = self.readOrthoDB(fullfilename)\n",
    "            elif myfile[2] == 4:\n",
    "                myproteins = self.readACBD5(fullfilename)\n",
    "            elif myfile[2] == 5:\n",
    "                myproteins = self.readUnirefByLine(fullfilename)\n",
    "            elif myfile[2] == 6:\n",
    "                myproteins = self.basicFileRead(fullfilename)\n",
    "            elif myfile[2] == 7:\n",
    "                myproteins = self.readAsOneLine(fullfilename)\n",
    "                \n",
    "            # 4 Nov 2018 replace target acids\n",
    "            self.replace_target_acids(myproteins)\n",
    "            \n",
    "            proteins.append([myfile,myproteins])\n",
    "        \n",
    "        return proteins\n",
    "        \n",
    "    '''\n",
    "    ****************************\n",
    "    * now files to read the individual files\n",
    "    ****************************\n",
    "    '''    \n",
    "        \n",
    "    def basicFileRead(self, fname):\n",
    "        \n",
    "        '''\n",
    "        ****************************\n",
    "        * 30 July 2019\n",
    "        * getting files with various headfers and forms, so creating a basic reader\n",
    "        * until it is clear a new more specific reader is worth creating\n",
    "        * approach is simple\" any line with '>' in is the start of a new protein\n",
    "        * so '>blah | ref |whatever' gets put whoesale in the proteinID, protein name is in variable standardname \n",
    "        *\n",
    "        * 29 Apr 2020\n",
    "        * slight adaptation for AAY0260_Table_s2\n",
    "        * this came in as a csv that I can convert to text. In doing this I have the first line > and then the code\n",
    "        * hence \n",
    "        *  >DHX34_47856039_11aa|upstream|\n",
    "        *  LCPSQRSLQMS\n",
    "        * So will look for a | for the ID then whatever follows minus the '|' is next\n",
    "        ****************************\n",
    "        '''\n",
    " \n",
    "        charset = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 =>_-'\n",
    "        standardname = 'not parsed - basic read'\n",
    "\n",
    "        f = open(fname, 'r')\n",
    "\n",
    "        proteins = []\n",
    "        thisprotein = [] \n",
    "\n",
    "        thissequence = ''\n",
    "        \n",
    "        for line in f:\n",
    "\n",
    "            if '>' in line:# got new protein\n",
    "\n",
    "                if len(thissequence) > 0:\n",
    "\n",
    "                    thisprotein = [[proteinID, proteinname]]\n",
    "                    thisprotein.append(thissequence)\n",
    "                    proteins.append(thisprotein)\n",
    "                    thissequence = ''\n",
    "                    \n",
    "                if '|' in line:\n",
    "                    \n",
    "                    # branch put in for AAY0260_Table_s2\n",
    "                    # or any others where I can create the file in a particular form\n",
    "                    \n",
    "                    newline = line[:].strip('\\n')\n",
    "                    \n",
    "                    \n",
    "                    IDidx = newline.index('|')\n",
    "                    proteinID = newline[1:IDidx] # assume '> is the 0th character at the start'\n",
    "                    proteinname = newline[IDidx:].strip('|')\n",
    "                    \n",
    "                else:\n",
    "\n",
    "                    proteinID = line[:]\n",
    "                    proteinname = standardname\n",
    "\n",
    "            else:\n",
    "                \n",
    "                ctr = 0\n",
    "                for char in line:\n",
    "                    ctr += 1\n",
    "                    if char in charset:\n",
    "                        thissequence += char     \n",
    "                        \n",
    "                thissequence = thissequence.strip('\\n\\r')\n",
    "\n",
    "        if len(thissequence) > 0:\n",
    "\n",
    "            thisprotein = [[proteinID, proteinname]]\n",
    "            thisprotein.append(thissequence)\n",
    "            proteins.append(thisprotein)\n",
    "\n",
    "        return proteins\n",
    "    \n",
    "    def readFASTA(self, fname):\n",
    "    \n",
    "        '''\n",
    "        ****************************\n",
    "        * reads a strand of protein in FASTA format\n",
    "        * in two parts: the first contains text and the second the code.\n",
    "        * the first part has a clear sequence:example is '>YHR073W OSH3 \n",
    "        * the YHR073W is a descriptive sequence. The second is a name\n",
    "        *\n",
    "        * data is in a list proteins\n",
    "        * each item is an individual protein sequence that contains two elements\n",
    "        * the first element is a list containing [7 letter id sequence, name, text]\n",
    "        * the second is a string of the code.\n",
    "        * \n",
    "        * August 2018 \n",
    "        * The previous way names were read in was wrong. They are not fixed length\n",
    "        * for example have these two\n",
    "        * 'YPR159C-A YPR159C-A SGDID:S000028725'\n",
    "        * 'YPR159W KRE6 SGDID:S000006363'\n",
    "        * need to keep going until a space, so have \n",
    "        *\n",
    "        * 2 Sep 2018\n",
    "        * seem to have screwed this up so trying to sort it. From human have \n",
    "        * sp|O95139|NDUB6_HUMAN NADH dehydrogenase [ubiquinone] 1 beta subcomplex subunit 6 OS=Homo sapiens OX=9606 GN=NDUFB6 PE=1 SV=3\n",
    "        * MTGYTPDEKLRLQQLRELRRRWLKDQELSPREPVLPPQKMGPMEKFWNKFLENKSPWRKM...\n",
    "        * becomes [['NDUB6','O95139'],['MTGYTPDEKLRLQQLRELRRRWLKDQELSPREPVLPPQKMGPMEKFWNKFLENKSPWRKM...\n",
    "        * \n",
    "        * 23 Sep 2018\n",
    "        * some files have an asterisk on the end of the acid chain - remove here.\n",
    "        ****************************\n",
    "        '''\n",
    "        \n",
    "        badchar = '*'\n",
    "\n",
    "        f = open(fname, 'r')\n",
    "\n",
    "        proteins = []\n",
    "        thisprotein = [] \n",
    "\n",
    "        for line in f:\n",
    "\n",
    "            if '>' in line:\n",
    "\n",
    "                # so have a new protein\n",
    "                if len(thisprotein) > 0:\n",
    "                    # If we have been building a protein, then we need to append the current protein to the list ...\n",
    "\n",
    "                    thisprotein.append(thissequence)\n",
    "                    proteins.append(thisprotein)\n",
    "                    \n",
    "                # now start a new protein\n",
    "\n",
    "                if '|' in line:\n",
    "                    \n",
    "                    idxID =line.index('|')\n",
    "                    proteinID = line[idxID + 1 : ] # 'O95139|NDUB6_HUMAN NADH deh'\n",
    "                    idxID = proteinID.index('|')\n",
    "                    proteinname = proteinID[:idxID].strip() # 'proteinname = 095139'\n",
    "            \n",
    "                    proteinID = proteinID[idxID + 1 : ]#'NDUB6_HUMAN NADH deh'\n",
    "                \n",
    "                    idxID = proteinID.index('_')\n",
    "                    if idxID < len(proteinID):\n",
    "\n",
    "                        proteinID = proteinID[:idxID]\n",
    "                        thisprotein = [[ proteinID , proteinname]]\n",
    "                    \n",
    "                    else:\n",
    "                        \n",
    "                        thisprotein = [[ 'unknown protein' , proteinname]]\n",
    "                        \n",
    "                else:\n",
    "                    \n",
    "                    thisdetail = []\n",
    "                    idxID = 1 + line.index('>')\n",
    "                    proteinID = ''\n",
    "                    while line[idxID] != ' ' and idxID<len(line):\n",
    "                        proteinID += line[idxID]\n",
    "                        idxID += 1\n",
    "\n",
    "                    proteinname = ''\n",
    "                    idxID += 1\n",
    "\n",
    "                    while line[idxID] !=' ' and idxID<len(line):\n",
    "                        proteinname += line[idxID]\n",
    "                        idxID += 1\n",
    "\n",
    "                    thisprotein = [[proteinID, proteinname, line[idxID:].strip('\\n')]]\n",
    "                \n",
    "                thissequence = ''\n",
    "            else:\n",
    "                thissequence += line.strip('\\n\\r')\n",
    "                while badchar in thissequence: # 23 Sep 2018 strip out *. usually on the end.\n",
    "                    idx = thissequence.index(badchar)\n",
    "                    thissequence = thissequence[:idx] + thissequence[idx + 1:]\n",
    "                \n",
    "        thisprotein.append(thissequence)\n",
    "        proteins.append(thisprotein)\n",
    "\n",
    "        return proteins\n",
    "    \n",
    "    def readUniRef50(self, fname):\n",
    "\n",
    "        '''\n",
    "        ********************************\n",
    "        * quite a lot of files cmoing in an unpleasant format with UniRef50 in the line\n",
    "        * this gives it a proper knife-and-forking to get the protein out\n",
    "        * but may lose some of thre associated detail and mangle some of the names.\n",
    "        * 4 Sep 2018\n",
    "        * not so easy. first line contains all the details, then the code follows\n",
    "        * introduce headerline as a differentiator\n",
    "        ********************************\n",
    "        '''\n",
    "\n",
    "        charset = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 =>_-'\n",
    "\n",
    "        f = open(fname, 'r')\n",
    "\n",
    "        myproteins = []\n",
    "        thisprotein = [] \n",
    "\n",
    "        thissequence = ''\n",
    "\n",
    "        for line in f:\n",
    "\n",
    "            if '>UniRef' in line:# got new protein\n",
    "                \n",
    "                if len(thissequence) > 0:\n",
    "                    \n",
    "                    thisprotein = [[proteinID, proteinname]]\n",
    "                    thisprotein.append(thissequence)\n",
    "                    myproteins.append(thisprotein)\n",
    "                    thissequence = ''\n",
    "                \n",
    "                headerline =''\n",
    "                for char in line:\n",
    "                    if char in charset:\n",
    "                        headerline += char \n",
    "                        \n",
    "                if 'RepID=' in headerline:\n",
    "                    idx1 = 6 + headerline.index('RepID=')\n",
    "                    if '_' in headerline[idx1:]:\n",
    "                        idx2 = idx1 + headerline[idx1:].index('_')\n",
    "                        proteinID = headerline[idx1 : idx2]\n",
    "                        proteinname= headerline[idx2+1:]\n",
    "                    else:\n",
    "                        proteinID = headerline[idx1 : ]\n",
    "                        proteinname = headerline[idx1:]\n",
    "                        \n",
    "                else:\n",
    "                    \n",
    "                    print ('non=spec code in Uniref')\n",
    "                    proteinname = 'unknown'\n",
    "                    proteinID = 'unknown'\n",
    "                    \n",
    "            else:\n",
    "                \n",
    "                for char in line:\n",
    "                    if char in charset:\n",
    "                        thissequence += char\n",
    "                    \n",
    "        if len(thissequence) > 0:\n",
    "                    \n",
    "            thisprotein = [[proteinID, proteinname]]\n",
    "            thisprotein.append(thissequence)\n",
    "            myproteins.append(thisprotein)\n",
    "            mycode = ''\n",
    "        \n",
    "        return myproteins\n",
    "    \n",
    "    def readOrthoDB(self, fname):\n",
    "\n",
    "        '''\n",
    "        ********************************\n",
    "        * 2 Sep 2018\n",
    "        * another format OrthoDB\n",
    "        * 226230:00043d {\"pub_gene_id\":\"J4TV27\", \"pub_og_id\":\"EOG092C5A9Y\", \"og_name\":\"Transcription factor Opi1\",\"level\":4751, \n",
    "        *  \"description\":\"Transcription factor Opi1\"} MSESQRLGLSEEEVEAAEVLGVLKQSCRQKSQRSENVSQGDRRLAGESSTTPLNILDRVSNKI\n",
    "        * becomes[['EOG092C5A9Y','J4TV27','Transcription factor Opi1'],'MSESQRLGLSEEEVEA']\n",
    "        ********************************\n",
    "        '''\n",
    "        \n",
    "        charset = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 =>_:{},*'\n",
    "\n",
    "        f = open(fname, 'r')\n",
    "\n",
    "        myproteins = []\n",
    "        thisprotein = [] \n",
    "        \n",
    "        thissequence = '' # need to do this for first protein\n",
    "        \n",
    "        for line in f:\n",
    "            \n",
    "            mycode = ''\n",
    "            \n",
    "            for char in line:\n",
    "                if char in charset:\n",
    "                    mycode += char\n",
    "\n",
    "            \n",
    "            if 'pub_gene_id' in mycode:\n",
    "                \n",
    "                # new protein\n",
    "                if len(thissequence) > 0:\n",
    "                    \n",
    "                    thisprotein = [[proteinID, proteinname]]\n",
    "                    thisprotein.append(thissequence)\n",
    "                    myproteins.append(thisprotein)\n",
    "                    \n",
    "                    thissequence = ''\n",
    "                    proteinname = proteinID = ''\n",
    "                    \n",
    "                idx1 = mycode.index('pub_gene_id')\n",
    "                substring = mycode[ idx1 : ]# \"pub_gene_id\":\"J4TV27\", \"pub_og_id\":\"EOG092C5A9Y\", \"\n",
    "                idx2 = substring.index(':')\n",
    "                substring = substring [ idx2 + 1 : ] # \"J4TV27\", \"pub_og_id\":\"EOG092C5A9Y\", \"\n",
    "                idx3 = substring.index(',')\n",
    "                proteinID = substring [ : idx3 ]\n",
    "\n",
    "                idx4 = substring.index(':')\n",
    "                substring = substring [ idx4 + 1 : ] # EOG092C5A9Y\", \"\n",
    "                idx5 = substring.index(',')\n",
    "                proteinname = substring [ : idx5 ]\n",
    "                               \n",
    "            else:\n",
    "\n",
    "                thissequence += mycode\n",
    "            \n",
    "        \n",
    "        if len(thissequence)>0:\n",
    "            \n",
    "            thisprotein = [[proteinID, proteinname]]\n",
    "            thisprotein.append(thissequence)\n",
    "            myproteins.append(thisprotein)\n",
    "\n",
    "        return myproteins\n",
    "\n",
    "    def readDashed(self, fname):\n",
    "        \n",
    "        '''\n",
    "        ********************************\n",
    "        * quite a lot of files coming in with an unpleasant format with UniRef50 in the line\n",
    "        * this gives it a proper knife-and-forking to get the protein out\n",
    "        * but may lose some of the associated detail and mangle some of the names.\n",
    "        *\n",
    "        * have input >tr|I7CJK2|I7CJK2_NATSJ Uncharacterized protein OS=Natrinema sp. (strain J7-2) GN=NJ7G_2660 PE=4 SV=1\n",
    "        *   LFDAQRTAVKQSQQLLKQGMT--------AQRTV-DTMVGTGLTGQESLQRYQLEVAQAA-------------THGTLSAMAA-----------------\n",
    "        *   --------------------------------------------------MLPG---DD--A--TEAHQAVDESFEQLKRTHAAVYDMLERDLEQGVDAT\n",
    "        * \n",
    "        * becomes\n",
    "        * tr|I7CJK2|I7CJK2_NATSJ Uncharacterized protein OS=Natrinema sp. (strain J7-2) GN=NJ7G_2660 PE=4 SV=1\n",
    "        *   LFDAQRTAVKQSQQLLKQGMTAQRTVDTMVGTGLTGQESLQRYQLEVAQAATHGTLSAMAAMLPGDDATEAHQAVDESFEQLKRTHAAVYDMLERDLEQGVDAT\n",
    "        * which we need to turn into [[I7CJK2,NATSJ],'LFDAQR...']\n",
    "        * having to use charactres culled from tests to identify ends of descriptions.\n",
    "        * should encourage getting rid if this format\n",
    "        ********************************\n",
    "        '''\n",
    "\n",
    "        charset = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 =>|_'\n",
    "        AAcharset = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'\n",
    "        \n",
    "        f = open(fname, 'r')\n",
    "\n",
    "        myproteins = []\n",
    "        thisprotein = [] \n",
    "\n",
    "        thissequence = ''\n",
    "        myheader = ''\n",
    "\n",
    "        for line in f:\n",
    "            \n",
    "            if '>' in line:\n",
    "                \n",
    "                if len(thissequence) > 0:#already have a protein assembled\n",
    "                    \n",
    "                    thisprotein = [[proteinID, proteinname]]\n",
    "                    thisprotein.append(thissequence)\n",
    "                    myproteins.append(thisprotein)\n",
    "                    thissequence = ''\n",
    "                \n",
    "                myheader = ''\n",
    "                for char in line:\n",
    "                    if char in charset:\n",
    "                        myheader += char\n",
    "                        \n",
    "                if '|' in myheader:\n",
    "\n",
    "                    idxu = 1 + myheader.index('|')  # 1 for length |\n",
    "                    myheader = myheader[idxu : ]\n",
    "\n",
    "                    if '|' in myheader:\n",
    "                        idxi = myheader.index('|') # before first space is the ID\n",
    "                        proteinID = myheader[ : idxi]\n",
    "                        myheader = myheader[ idxi + 1 :]\n",
    "\n",
    "                    if '_' in myheader:\n",
    "                        idxn = myheader.index('_') # before second space is the name\n",
    "                        myheader = myheader[ idxn + 1:]\n",
    "\n",
    "                    if ' ' in myheader:\n",
    "                        idxn = myheader.index(' ') # before second space is the name\n",
    "                        proteinname = myheader[:idxn]\n",
    "                        myheader = myheader[ idxn + 1:]\n",
    "                        \n",
    "            else:\n",
    "                \n",
    "                for char in line:\n",
    "                    if char in AAcharset:\n",
    "                        thissequence += char\n",
    "                        \n",
    "        if len(thissequence) > 0: # have the last protein assembled and need to add.\n",
    "                    \n",
    "            thisprotein = [[proteinID, proteinname]]\n",
    "            thisprotein.append(thissequence)\n",
    "            myproteins.append(thisprotein)\n",
    "\n",
    " \n",
    "        return myproteins\n",
    "\n",
    "    def readACBD5(self, fname):\n",
    "    \n",
    "        '''\n",
    "        ****************************\n",
    "        * format ACDB5 called for files received from another source\n",
    "        * which are two files with ACBD5 in the title\n",
    "        * very simple format: first line is just the protein '>A0FKI7\n",
    "        * then the amino acid code\n",
    "        * 16 Sep 2018\n",
    "        *\n",
    "        * 26 July 2019\n",
    "        * using this fior a 'fasta' file 'PI_genemodels_prot_jan19_2017_fasta'\n",
    "        * this finishes with an asterisk so updating to strip trailing *\n",
    "        ****************************\n",
    "        '''\n",
    "        \n",
    "        badchar = '*' # Added 26 July see above.\n",
    "            \n",
    "        f = open(fname, 'r')\n",
    "\n",
    "        proteins = []\n",
    "        thisprotein = [] \n",
    "\n",
    "        for line in f:\n",
    "\n",
    "            if '>' in line:\n",
    "\n",
    "                # so have a new protein\n",
    "                # we need to complete the current protein and add to the list ...\n",
    "                \n",
    "                if len(thisprotein) > 0:\n",
    "\n",
    "                    thisprotein.append(thissequence)\n",
    "                    proteins.append(thisprotein)\n",
    "                    \n",
    "                # now start a new protein\n",
    "                # format is simple - '>' then the protein name eg >A0FKI7\n",
    "                \n",
    "                idxID =line.index('>')\n",
    "                proteinID = line[idxID + 1 : ].strip() # 'A0FKI7' - that's all there is\n",
    "                thisprotein = [[ proteinID , proteinID]]\n",
    "                \n",
    "                thissequence = ''\n",
    "                        \n",
    "            else:\n",
    "                \n",
    "                thissequence += line.strip('\\n\\r')\n",
    "                # 26 July 2019 strip trailing *\n",
    "                while badchar in thissequence: # 23 Sep 2018 strip out *. usually on the end.\n",
    "                    idx = thissequence.index(badchar)\n",
    "                    thissequence = thissequence[:idx] + thissequence[idx + 1:]\n",
    "                \n",
    "        #add final protein\n",
    "        thisprotein.append(thissequence)\n",
    "        proteins.append(thisprotein)\n",
    "\n",
    "        return proteins\n",
    "    \n",
    "    def readUnirefByLine(self, fname):\n",
    "        \n",
    "        '''\n",
    "        ****************************\n",
    "        * reads a strand of protein in adjusted Uniref Format\n",
    "        * so have two lines\n",
    "        * >UniRef90_A0A022Q2S6 Uncharacterized protein n=1 Tax=Erythranthe guttata TaxID=4155 RepID=A0A022Q2S6_ERYGU\n",
    "        * MGRLQLFDDGDDSPNGGEDDLSKIEINHEFAKRYEYNKKREELQKYEELKKQGRVD...\n",
    "        * all comes in one line separated by \\r\n",
    "        ****************************\n",
    "        '''\n",
    "        \n",
    "        proteins = []\n",
    "\n",
    "        f = open(fname, 'r')\n",
    "        \n",
    "        for line in f:\n",
    "    \n",
    "            mychars = line[:]\n",
    "    \n",
    "            ctr = 0\n",
    "\n",
    "            while '>' in mychars and ctr<1000:\n",
    "        \n",
    "                ctr += 1\n",
    "\n",
    "                idx = mychars.index('>')\n",
    "\n",
    "                thisprotein = mychars[:idx]\n",
    "                if len(thisprotein) > 0:\n",
    "\n",
    "                    idx1 = thisprotein.index('UniRef90_')\n",
    "                    thisprotein = thisprotein[idx1 + 9:]\n",
    "\n",
    "                    idx2 = thisprotein.index(' ')\n",
    "                    thisname = thisprotein[:idx2]\n",
    "                    thisprotein = thisprotein[idx2:]\n",
    "\n",
    "                    idx3 = thisprotein.index('TaxID=')\n",
    "                    thisprotein = thisprotein[idx3 + 6:]\n",
    "\n",
    "                    idx4 = thisprotein.index(' ')\n",
    "                    thisref = thisprotein[:idx4]\n",
    "                    thisprotein = thisprotein[idx4:] \n",
    "\n",
    "                    idx5 = thisprotein.index('_')\n",
    "\n",
    "                    thisprotein = thisprotein[idx5 + 7:] \n",
    "\n",
    "                    proteins.append([[thisname,thisref],thisprotein.strip('\\r')])\n",
    "\n",
    "\n",
    "                mychars = mychars[idx + 1:]\n",
    "\n",
    "        \n",
    "        idx1 = mychars.index('UniRef90_')\n",
    "        thisprotein = mychars[idx1 + 9:]\n",
    "\n",
    "        thisname = thisprotein[:6]\n",
    "\n",
    "        idx3 = thisprotein.index('TaxID=')\n",
    "        thisprotein = thisprotein[idx3 + 6:]\n",
    "\n",
    "        idx4 = thisprotein.index(' ')\n",
    "        thisref = thisprotein[:idx4]\n",
    "        thisprotein = thisprotein[idx4:] \n",
    "\n",
    "        idx5 = thisprotein.index('_')\n",
    "\n",
    "        thisprotein = thisprotein[idx5 + 7:] \n",
    "        proteins.append([[thisname,thisref],thisprotein.strip('\\r')])\n",
    "        \n",
    "        return proteins\n",
    "    \n",
    "    def readAsOneLine(self, fname):\n",
    "        \n",
    "        '''\n",
    "        ****************************\n",
    "        * 9 Aug 2019\n",
    "        * some files seem to come in as one line\n",
    "        * need to unpack them in a general way\n",
    "        * this is not a universal reader and some that come in as one line end up improperly parsed\n",
    "        * with no name in the header sectionand a load of headerinfo in the amino acid section\n",
    "        ****************************\n",
    "        '''\n",
    "\n",
    "        charset = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 =>_-[]'\n",
    "        endchar = ']'\n",
    "        startchar = '>'\n",
    "\n",
    "        proteins = []\n",
    "        myheader = ''\n",
    "        myaacode = ''\n",
    "        \n",
    "        f = open(fname, 'r')\n",
    "\n",
    "        for line in f:\n",
    "\n",
    "            mychars = line\n",
    "\n",
    "            recog = ''\n",
    "\n",
    "            for char in mychars:\n",
    "\n",
    "                if char == startchar:\n",
    "\n",
    "                    if len(recog) > 0:\n",
    "                        myaacode = recog\n",
    "                        thisprotein = []\n",
    "                        thisprotein.append([myheader,'(basic - read)'])\n",
    "                        thisprotein.append(myaacode)\n",
    "                        proteins.append(thisprotein)\n",
    "                        recog = ''\n",
    "\n",
    "                elif char in charset:\n",
    "                    recog += char\n",
    "\n",
    "                if char == endchar:\n",
    "                    if len(recog) > 0:\n",
    "                        myheader = recog\n",
    "                        recog = ''\n",
    "\n",
    "            if len(recog) > 0:\n",
    "                thisprotein = []\n",
    "                thisprotein.append([myheader,'(one line read)'])\n",
    "                thisprotein.append(recog)\n",
    "                proteins.append(thisprotein)\n",
    "\n",
    "        return proteins \n",
    "    \n",
    "    def replace_target_acids(self, proteins):\n",
    "        '''\n",
    "        ****************************\n",
    "        * 4 Nov 2018\n",
    "        * replaces specific amino acids with other amino acids in the proteome.\n",
    "        * need arose to replace 'O' with 'X' in Methanosarcina Mazei proteome\n",
    "        * but make a general case\n",
    "        * so creat a list of pairs, with the first item being replaced with second item\n",
    "        ****************************\n",
    "        '''\n",
    "        \n",
    "        targetacids = []\n",
    "        targetacids.append(['O','X'])\n",
    "    \n",
    "        for protein in proteins:\n",
    "            ctr = 0\n",
    "            for target in targetacids:\n",
    "                while target[0] in protein[1]:\n",
    "                    idx = protein[1].index(target[0])\n",
    "                    protein[1] = protein[1][:idx] + target[1] + protein[1][idx + 1:]\n",
    "                    ctr += 1\n",
    "            if ctr > 0:\n",
    "                # should log this, but currently no log file available in this module\n",
    "                if ctr == 1:\n",
    "                    print ('replaced 1 instance of a rogue amino acid in ' + protein[0][0])\n",
    "                else:\n",
    "                    print ('replaced ' + str(ctr) + ' instances of rogue amino acids in ' + protein[0][0])\n",
    "            \n",
    "    '''\n",
    "****************************\n",
    "* end of the proteome reader class\n",
    "****************************\n",
    "'''\n",
    "\n",
    "# the line below added to enable check that corrections or additions have been accessed. Suggest incrementing last \n",
    "# number each time a change is made.\n",
    "print ('Core GH 22 04')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
